<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lab 6: Applying Functional Programming with Purrr to Models – R for Bio Data Science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./lab07.html" rel="next">
<link href="./lab05.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-bdef442aaf320a139b1684e38e335c71.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./lab00.html">Course Labs</a></li><li class="breadcrumb-item"><a href="./lab06.html">Lab 6: Applying Functional Programming with Purrr to Models</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">R for Bio Data Science</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome to R for Bio Data Science</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prologue.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prologue</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Getting Started</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./lab00.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Course Labs</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 1: Course Intro &amp; the Very Basics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 2: Data Visualisation I</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 3: Data Visualisation II</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 4: Data Wrangling I</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 5: Data Wrangling II</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab06.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Lab 6: Applying Functional Programming with Purrr to Models</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 7: Collaborative Bio Data Science using GitHub via RStudio</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 8: Creating a Simple R-package</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 9 Creating a Simple Shiny Application</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Lab 10 Project Startup &amp; Industry Talks</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./primers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Primers</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_linear_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Primer on Linear Models in <code>R</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_logistic_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Primer on Logistic Regression in <code>R</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Primer on Functions in <code>R</code></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_r_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Primer on R package development</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_paths_and_projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Paths &amp; Projects</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./primer_variable_assignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variable Assignment in <code>R</code></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./guides.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Guides</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./guide_rstudio_server.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Guide for Cloud server and the RStudio IDE</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./guide_working_locally.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Guide for Working Locally</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./guide_building_r4ds2e_locally.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Guide for Building the R4DS2e Book Locally</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./code_styling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code Styling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./external_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">External Resources</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./course_elements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Course Elements</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ai_in_r4bds.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Use of AI in the R for Bio Data Science Course</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assignments</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_description.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project Description</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_faq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project FAQ</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_checklist.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project Checklist</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exam.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exam</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#packages" id="toc-packages" class="nav-link active" data-scroll-target="#packages">Package(s)</a></li>
  <li><a href="#schedule" id="toc-schedule" class="nav-link" data-scroll-target="#schedule">Schedule</a></li>
  <li><a href="#learning-materials" id="toc-learning-materials" class="nav-link" data-scroll-target="#learning-materials">Learning Materials</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#sec-exercises" id="toc-sec-exercises" class="nav-link" data-scroll-target="#sec-exercises">Exercises</a></li>
  <li><a href="#throwback" id="toc-throwback" class="nav-link" data-scroll-target="#throwback">Throwback…</a></li>
  <li><a href="#prologue" id="toc-prologue" class="nav-link" data-scroll-target="#prologue">Prologue</a>
  <ul class="collapse">
  <li><a href="#getting-started" id="toc-getting-started" class="nav-link" data-scroll-target="#getting-started">Getting Started</a></li>
  </ul></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load Data</a></li>
  <li><a href="#clean-data" id="toc-clean-data" class="nav-link" data-scroll-target="#clean-data">Clean Data</a></li>
  <li><a href="#augment-data" id="toc-augment-data" class="nav-link" data-scroll-target="#augment-data">Augment Data</a></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">Analysis</a>
  <ul class="collapse">
  <li><a href="#one-gene-one-model" id="toc-one-gene-one-model" class="nav-link" data-scroll-target="#one-gene-one-model">One Gene, one model</a></li>
  <li><a href="#all-the-genes-all-the-models" id="toc-all-the-genes-all-the-models" class="nav-link" data-scroll-target="#all-the-genes-all-the-models">All the Genes, all the models</a></li>
  <li><a href="#models-models-everywhere" id="toc-models-models-everywhere" class="nav-link" data-scroll-target="#models-models-everywhere">Models, models everywhere…</a></li>
  </ul></li>
  <li><a href="#group-assignment" id="toc-group-assignment" class="nav-link" data-scroll-target="#group-assignment"><span style="color: red;">GROUP ASSIGNMENT</span></a></li>
  <li><a href="#optional" id="toc-optional" class="nav-link" data-scroll-target="#optional">Optional</a>
  <ul class="collapse">
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#k-means" id="toc-k-means" class="nav-link" data-scroll-target="#k-means">K-means</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./lab00.html">Course Labs</a></li><li class="breadcrumb-item"><a href="./lab06.html">Lab 6: Applying Functional Programming with Purrr to Models</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Lab 6: Applying Functional Programming with Purrr to Models</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="packages" class="level2">
<h2 class="anchored" data-anchor-id="packages">Package(s)</h2>
<ul>
<li><a href="https://broom.tidymodels.org">broom</a></li>
<li><a href="https://purrr.tidyverse.org">purrr</a></li>
</ul>
</section>
<section id="schedule" class="level2">
<h2 class="anchored" data-anchor-id="schedule">Schedule</h2>
<ul>
<li>08.00 - 08.45: <a href="https://raw.githack.com/r4bds/r4bds.github.io/main/recap_lab06.html">Recap of Lab 5</a></li>
<li>08.45 - 09.00: <a href="https://raw.githack.com/r4bds/r4bds.github.io/main/lecture_lab06.html">Lecture</a></li>
<li>09.00 - 09.15: Break</li>
<li>09.00 - 12.00: <a href="#sec-exercises">Exercises</a></li>
</ul>
</section>
<section id="learning-materials" class="level2">
<h2 class="anchored" data-anchor-id="learning-materials">Learning Materials</h2>
<p><em>Please prepare the following materials:</em></p>
<ul>
<li><strong>Important:</strong> Please fill in this <a href="./getting_started.html#sec-midwayeval">brief midway evaluation questionnaire</a> (est. ~2 min)</li>
<li>Book (Note, this is intentionally 1.ed.): <a href="https://r4ds.had.co.nz/model-intro.html">R4DS: Chapter 22: Introduction</a></li>
<li>Book (Note, this is intentionally 1.ed.): <a href="https://r4ds.had.co.nz/model-basics.html">Chapter 23: Model Basics</a></li>
<li>Book (Note, this is intentionally 1.ed.): <a href="https://r4ds.had.co.nz/model-building.html">Chapter 24: Model Building</a></li>
<li>Book (Note, this is intentionally 1.ed.): <a href="https://r4ds.had.co.nz/many-models.html">Chapter 25: Many models</a></li>
<li>Video: <a href="https://www.youtube.com/watch?v=7VGPUBWGv6g">Broom: Converting Statistical Models to Tidy Data Frames</a></li>
<li>Video: <a href="https://www.youtube.com/watch?v=gLZyvY5_kWQ">Alex Hayes | Solving the model representation problem with broom | RStudio (2019)</a></li>
<li>Video: <a href="https://www.youtube.com/watch?v=bzUmK0Y07ck">“The Joy of Functional Programming (for Data Science)” with Hadley Wickham</a></li>
<li>Optional: If you are completely new to statistical modelling, then <a href="./primer_linear_models.html">click here for a primer</a></li>
</ul>
<p><em>Unless explicitly stated, do not do the per-chapter exercises in the R4DS2e book</em></p>
</section>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<p><em>A student who has met the objectives of the session will be able to:</em></p>
<ul>
<li>Fit a simple linear model and interpret model parameters</li>
<li>Understand and apply simple purrr-functions for element-wise function application</li>
<li>Understand and apply grouped supervised models to form nested model objects</li>
<li>Understand and apply the broom-functions for tidying various model objects</li>
<li>Optional LO: Perform a basic principal component analysis for dimension reduction of high dimensional data</li>
<li>Optional LO: Perform a basic unsupervised k-means clustering of high dimensional data</li>
</ul>
</section>
<section id="sec-exercises" class="level2">
<h2 class="anchored" data-anchor-id="sec-exercises">Exercises</h2>
</section>
<section id="throwback" class="level2">
<h2 class="anchored" data-anchor-id="throwback">Throwback…</h2>
<p>Using the <code>tibble()</code> function, re-create the data from this visualisation and then create your version of how this data could be visualised in a more informative manner:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/bad_viz_15.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
<p>Also… If you still need to be convinced of the flexibility of ggplot, try running this code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="sc">-</span><span class="dv">3</span>,</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">to =</span> <span class="dv">3</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">by =</span> <span class="fl">0.01</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">expand.grid</span>(<span class="at">x =</span> xy,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">y =</span> xy) <span class="sc">|&gt;</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">mapping =</span> <span class="fu">aes</span>(</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> (<span class="dv">1</span> <span class="sc">-</span> x <span class="sc">-</span> <span class="fu">sin</span>(y<span class="sc">^</span><span class="dv">2</span>)),</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> (<span class="dv">1</span> <span class="sc">+</span> y <span class="sc">-</span> <span class="fu">cos</span>(x<span class="sc">^</span><span class="dv">2</span>)))) <span class="sc">+</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> <span class="dv">20</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_polar</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you are curious about what is going on here, try googling “Generative art”… Anyhoo… Let us move on…</p>
</section>
<section id="prologue" class="level2">
<h2 class="anchored" data-anchor-id="prologue">Prologue</h2>
<p>So far, we have worked on</p>
<ul>
<li>Lab 2: Genomics, the <code>gravier</code> data from <span class="citation" data-cites="Gravier2010">Gravier et al. (<a href="references.html#ref-Gravier2010" role="doc-biblioref">2010</a>)</span>.</li>
<li>Lab 3: Metagenomics, the <code>pitlatrine</code> data from <span class="citation" data-cites="Torondel2016">Torondel et al. (<a href="references.html#ref-Torondel2016" role="doc-biblioref">2016</a>)</span>.</li>
<li>Lab 4: Clinical data, the <code>diabetes</code> data from <span class="citation" data-cites="Willems1997">Willems et al. (<a href="references.html#ref-Willems1997" role="doc-biblioref">1997</a>)</span> and <span class="citation" data-cites="Schorling1997">Schorling et al. (<a href="references.html#ref-Schorling1997" role="doc-biblioref">1997</a>)</span>.</li>
<li>Lab 5: High throughput immunoinformatics data, the <code>SARS-CoV-2</code> data from <span class="citation" data-cites="Nolan2020">Nolan et al. (<a href="references.html#ref-Nolan2020" role="doc-biblioref">2020</a>)</span>.</li>
</ul>
<section id="getting-started" class="level3">
<h3 class="anchored" data-anchor-id="getting-started">Getting Started</h3>
<p><em>First, make sure to read and discuss the feedback you got from last week’s assignment!</em></p>
<p>Then, once again go to the <a href="https://teaching.healthtech.dtu.dk/22100/rstudio.php">R for Bio Data Science RStudio Cloud Server</a></p>
<p>Create a <strong>new</strong> Quarto document and remember to save it</p>
<p>Today, we will re-examine the data acquired in Lab 2 to further explore the research question:</p>
<ul>
<li><strong>What genes are significantly up- or down-regulated between the patients with/without early metastasis?“</strong></li>
</ul>
</section>
</section>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load Data</h2>
<p><strong>Before we do this, go to the console and enter the command: <code>rm(list = ls())</code></strong></p>
<p><em>This will clear your environment from old object from previous labs</em></p>
<p>Good now that we have a clean slate - Recall, so far we have adjusted the chunk setting to <code>#| eval: true</code> first time we get some data from an online repository and then subsequently set <code>#| eval: false</code>.</p>
<p>Let’s do that in a bit nicer way.</p>
<ul>
<li><strong>T1: Create a new “Load Data” header in your document and add the below chunk:</strong></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>raw_dir <span class="ot">&lt;-</span> <span class="st">"data/_raw/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data_file <span class="ot">&lt;-</span> <span class="st">"gravier.RData"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>data_loc <span class="ot">&lt;-</span> <span class="st">"https://github.com/ramhiser/datamicroarray/raw/master/data/"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="sc">!</span><span class="fu">dir.exists</span>(raw_dir) ){</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="at">path =</span> raw_dir)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( <span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">str_c</span>(raw_dir, data_file)) ){</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">url =</span> <span class="fu">str_c</span>(data_loc, data_file),</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">destfile =</span> <span class="fu">str_c</span>(raw_dir, data_file))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="at">file =</span> <span class="fu">str_c</span>(raw_dir, data_file))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><strong>Q1: In your group, discuss, what is going on here? Make sure you follow the difference between the first time you run this and the second!</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<p>We have used the <code>str_c()</code> function before, try running e.g.:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="st">"a"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="st">"b"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str_c</span>(x, y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the functions <code>dir.exists()</code> and <code>file.exists()</code>, the hint is in the title, they return logicals. To get a better understanding of this, try running e.g.:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>x <span class="sc">==</span> <span class="dv">2</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>( x <span class="sc">==</span> <span class="dv">2</span> ){</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Yes!"</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then change <code>x == 2</code> to <code>x != 2</code>, re-run the code and see what happens</p>
</details>
</section>
<section id="clean-data" class="level2">
<h2 class="anchored" data-anchor-id="clean-data">Clean Data</h2>
<p>The next step is to clean up the data:</p>
<ol type="1">
<li>Use the <code>ls()</code> function to see what objects you have in your environment</li>
<li>Use the <code>str()</code> function on the gravier data you retrieved to answer:</li>
</ol>
<ul>
<li><p><strong>Q2: Discuss in your group if this is tidy data?</strong></p></li>
<li><p><strong>T2: Create a new “Clean Data” header in your document and add the below chunk:</strong></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>gravier_clean <span class="ot">&lt;-</span> gravier <span class="sc">|&gt;</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>() <span class="sc">|&gt;</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p><strong>Q3: Discuss in your group if this is tidy data?</strong></p></li>
<li><p><strong>Q4: In fact, specifically state what are the “rules” for tidy data?</strong></p></li>
<li><p><strong>Q5: In your group, discuss why <code>bind_cols</code> can by <em>very very</em> dangerous to use?</strong></p></li>
</ul>
<p>Now, moving on, let’s write the clean data to disk:</p>
<ul>
<li><strong>T3: In your “Clean Data”-section, add a new chunk, where you write a tab-separated-values gzipped (i.e.&nbsp;compressed) file called “02_gravier_clean” (with the correct file type specification) into your “data”-folder</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Just as you can <code>read</code> a file, you can of course also <code>write</code> a file. Note the file type we want to write here is tab-separated-values. If you in the console type e.g.&nbsp;<code>readr::wr</code> and then hit the Tab key, you will see the different functions for writing different file types. We previously did a trick to automatically gzip (compress) files?</em>
</details>
</section>
<section id="augment-data" class="level2">
<h2 class="anchored" data-anchor-id="augment-data">Augment Data</h2>
<ul>
<li><strong>T4: Create a new “Augment Data” header in your document and add the below chunk:</strong></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug <span class="ot">&lt;-</span> gravier_clean <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">y =</span> <span class="fu">case_when</span>(y <span class="sc">==</span> <span class="st">"poor"</span> <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                       y <span class="sc">==</span> <span class="st">"good"</span> <span class="sc">~</span> <span class="dv">0</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(<span class="at">early_metastasis =</span> y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p><strong>Q6: In your group, discuss, what each step of the above work flow does, i.e.&nbsp;what are the specifics of the <code>dplyr</code> pipeline?</strong></p></li>
<li><p><strong>T5: In your “Augment Data”-section, add a new chunk, where you write a tab-separated-values gzipped (i.e.&nbsp;compressed) file called “03_gravier_clean_aug” (with the correct file type specification) into your “data” folder</strong></p></li>
</ul>
</section>
<section id="analysis" class="level2">
<h2 class="anchored" data-anchor-id="analysis">Analysis</h2>
<section id="one-gene-one-model" class="level3">
<h3 class="anchored" data-anchor-id="one-gene-one-model">One Gene, one model</h3>
<ul>
<li><strong>T6: Create a new “Analysis” header in your document</strong></li>
</ul>
<p>Recall, in the second lab, we were looking at “our favourite gene”. In the following either look back to what was your favourite gene or choose a new 🤷️</p>
<p>Let’s fit our first model! If the concept of models and linear regression is unfamiliar, consider checking out the <a href="./primer_linear_models.html">primer on linear models in R</a> before proceeding.</p>
<ul>
<li><strong>T7: Use the <code>lm()</code> function to create your first model and save it to a new variable e.g.&nbsp;“my_first_model”</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Use the formula <code>my_favourite_gene ~ early_metastasis</code> and remember when you pipe into the <code>lm()</code> function, you have to specify <code>data = _</code></em>
</details>
<ul>
<li><strong>Q7: What are your coefficients? Mine are:</strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     (Intercept) early_metastasis 
     -0.01616011      -0.03426164 </code></pre>
</div>
</div>
<ul>
<li><p><strong>T8: Use the <code>group_by()</code> <span class="math inline">\(\rightarrow\)</span> <code>summarise()</code> workflow to calculate the mean values of the gene expression for your favourite gene stratified on <code>early_metastasis</code></strong></p></li>
<li><p><strong>Q8: What are your mean values? Mine are:</strong></p></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  early_metastasis      mu
             &lt;dbl&gt;   &lt;dbl&gt;
1                0 -0.0162
2                1 -0.0504</code></pre>
</div>
</div>
<ul>
<li><strong>Q9: Discuss in your group: How are your coefficients related to your mean expression values?</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Recall, we have two terms here, <code>intercept</code> and <code>slope</code>. The intercept is the <code>y</code> value at <code>x = 0</code> and the <code>slope</code> is the change in <code>y</code> value, for one unit change in <code>x</code></em>
</details>
<ul>
<li><strong>Q10: Discuss in your group: Is your gene up- or down-regulated from <code>early_metastasis = 0</code> to <code>early_metastasis = 1</code> and use the <code>summary()</code> function to see if is it statistically significant at a level of significance of <span class="math inline">\(\alpha = 0.05\)</span>?</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Try to run <code>my_first_model |&gt; summary()</code> and look at the estimate for <code>early_metastasis</code>, i.e.&nbsp;the slope and also see if you can find the <code>p-value</code> somewhere in this summary…</em>
</details>
<p>Excellent! Now we have a feeling for working with the <code>lm()</code> functions and a basic understanding of the meaning of the coefficients, when we are using linear regression to model a binary outcome. The reason that we use a linear regression model in this case is, that for these exercises, we want to investigate the relationship between variables rather than obtaining probability predictions, i.e.</p>
<ul>
<li><strong>What genes are significantly up-/down-regulated between the patients with- and without early metastasis?</strong></li>
</ul>
<p><em>So, without further ado, let’s dive in!</em></p>
</section>
<section id="all-the-genes-all-the-models" class="level3">
<h3 class="anchored" data-anchor-id="all-the-genes-all-the-models">All the Genes, all the models</h3>
<p>First, the recent couple of years have seen an immense development in unifying the modelling interface in <code>R</code>, which is notoriously inconsistent. You may be familiar with the <code>caret</code> package, the developer of which has created <a href="https://www.tidymodels.org">tidymodels</a>, (which I really wish we had time to explore in details). In the following we will work with some of the principles for tidying model object using <code>broom</code>, having object nested in tibbles and working with these using <code>purrr</code>.</p>
<p>Now, you saw above how we could fit <em>one</em> model for <em>one</em> gene. So, we could repeat the procedure you worked through for each gene, but first consider:</p>
<ul>
<li><strong>Q11: How many genes are there in the gravier data set?</strong></li>
</ul>
<p>Now, we just have to make one model for each gene!</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/statistical_models_everywhere.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
<p>Honestly, <strong>let’s not!</strong> Also, recall: <em>“We don’t loop, we func!”</em>, so let’s see how that would work.</p>
</section>
<section id="models-models-everywhere" class="level3">
<h3 class="anchored" data-anchor-id="models-models-everywhere">Models, models everywhere…</h3>
<p>In principle, you could overflow your environment with individual model objects, but that would require <strong>a lot</strong> of code-lines and <strong>a lot</strong> of hard-coding. But let us instead see if we can come up with something just a tad more clever.</p>
<section id="preparing-the-data" class="level4">
<h4 class="anchored" data-anchor-id="preparing-the-data">Preparing the Data</h4>
<p>First:</p>
<ul>
<li><strong>Q12: Discuss in your group, if the <code>gravier_clean_aug</code> is a “wide” or a “long” dataset?</strong></li>
</ul>
<p>Once you have agreed upon and understood why, this is a <em>wide</em> data set, proceed and:</p>
<ul>
<li><strong>T9: Create this long version of your <code>gravier_clean_aug</code> data and save it in <code>gravier_clean_aug_long</code></strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 488,040 × 3
   early_metastasis gene   log2_expr_level
              &lt;dbl&gt; &lt;chr&gt;            &lt;dbl&gt;
 1                0 g2E09         -0.00144
 2                0 g7F07         -0.00144
 3                0 g1A01         -0.0831 
 4                0 g3C09         -0.0475 
 5                0 g3H08          0.0158 
 6                0 g1A08         -0.0336 
 7                0 g1B01         -0.136  
 8                0 g1int1         0.0180 
 9                0 g1E11          0.0257 
10                0 g8G02          0.00720
# ℹ 488,030 more rows</code></pre>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Recall which function took a dataset from “wide” to “long” format and also the three parameters of interest for that function is <code>cols</code>, <code>names_to</code> and <code>values_to</code>. If you look at my example, you will see that the columns we have pivoted have types <code>&lt;chr&gt;</code> and <code>&lt;dbl&gt;</code>, these will map to the <code>names_to</code> and <code>values_to</code> parameters respectively. The <code>cols</code> can be used with the so-called “helper-functions”, that we have previously talked about. If you look at the gene-columns in the data, all the genes starts with a “g”, so we can use the helper functions <code>starts_with()</code> and then give the argument “g” to the <code>match</code> parameter</em>
</details>
<p>The reason we want the “long” version of the data set is, that now we have ALL the genes defined in ONE <code>gene</code> column. This means that we can use <code>group_by()</code> to work <em>per</em> gene without looping (<em>Recall: Don’t loop, only func!</em>).</p>
<p>Now:</p>
<ul>
<li><strong>T10: Create a <code>dplyr</code> pipeline, use the <code>group_by()</code> function to group your <code>gravier_clean_aug_long</code> dataset by <code>gene</code> and then add the <code>nest()</code> and <code>ungroup()</code> functions to your pipeline</strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,905 × 2
   gene   data              
   &lt;chr&gt;  &lt;list&gt;            
 1 g2E09  &lt;tibble [168 × 2]&gt;
 2 g7F07  &lt;tibble [168 × 2]&gt;
 3 g1A01  &lt;tibble [168 × 2]&gt;
 4 g3C09  &lt;tibble [168 × 2]&gt;
 5 g3H08  &lt;tibble [168 × 2]&gt;
 6 g1A08  &lt;tibble [168 × 2]&gt;
 7 g1B01  &lt;tibble [168 × 2]&gt;
 8 g1int1 &lt;tibble [168 × 2]&gt;
 9 g1E11  &lt;tibble [168 × 2]&gt;
10 g8G02  &lt;tibble [168 × 2]&gt;
# ℹ 2,895 more rows</code></pre>
</div>
</div>
<p>Note, this is conceptually a <strong>super-tricky data structure!</strong></p>
<ul>
<li><strong>Q13: Discuss in your group, what happened to the data?</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<p><em>Try to run each of the following code chunks and examine what happens at each step:</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested <span class="sc">|&gt;</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"g2E09"</span>) <span class="co"># Replace "g2E09" with whatever was YOUR favourite gene!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested <span class="sc">|&gt;</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"g2E09"</span>) <span class="sc">|&gt;</span> <span class="co"># Replace "g2E09" with whatever was YOUR favourite gene!</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<ul>
<li><strong>Q14: Moreover, discuss in your group, what does <code>&lt;tibble [168 × 2]&gt;</code> mean?</strong></li>
</ul>
<p>Now, <strong>if</strong> you’re experiencing, that the server seems slow, consider if you want to proceed now with ALL the genes or just a subset. If you just want a subset, then use <code>sample_n()</code> to randomly select e.g.&nbsp;100 genes for further analysis. Remember you may want to use the <code>set.seed()</code> function to create a reproducible work flow even when sampling.</p>
</section>
<section id="fitting-models" class="level4">
<h4 class="anchored" data-anchor-id="fitting-models">Fitting Models</h4>
<p>Now, recall our research question:</p>
<ul>
<li><strong>What genes are significantly up-/down-regulated between the patients with- and without early metastasis?</strong></li>
</ul>
<p>To investigate this, we want to fit a linear model to each gene, i.e.&nbsp;as you did initially for your favourite gene, we want to do in a clever way <em>per</em> gene for ALL genes.</p>
<ul>
<li><strong>T11: Use the <code>group_by()</code> function to let <code>R</code> know, that we want to work <em>per</em> gene</strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,905 × 2
# Groups:   gene [2,905]
   gene   data              
   &lt;chr&gt;  &lt;list&gt;            
 1 g2E09  &lt;tibble [168 × 2]&gt;
 2 g7F07  &lt;tibble [168 × 2]&gt;
 3 g1A01  &lt;tibble [168 × 2]&gt;
 4 g3C09  &lt;tibble [168 × 2]&gt;
 5 g3H08  &lt;tibble [168 × 2]&gt;
 6 g1A08  &lt;tibble [168 × 2]&gt;
 7 g1B01  &lt;tibble [168 × 2]&gt;
 8 g1int1 &lt;tibble [168 × 2]&gt;
 9 g1E11  &lt;tibble [168 × 2]&gt;
10 g8G02  &lt;tibble [168 × 2]&gt;
# ℹ 2,895 more rows</code></pre>
</div>
</div>
<ul>
<li><strong>T12: Then using the <code>map()</code>-function, add a new line to your pipeline, where you add a new variable <code>model_object</code> to your <code>gravier_clean_aug_long_nested</code> dataset, which <code>R</code> will compute <em>per</em> gene</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<p><em>To do this you will need the following syntax and then wrap that inside the relevant tidyverse-function for creating a new variable:</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>model_object <span class="ot">=</span> <span class="fu">map</span>(<span class="at">.x =</span> data,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.f =</span> <span class="sc">~</span><span class="fu">lm</span>(<span class="at">formula =</span> log2_expr_level <span class="sc">~</span> early_metastasis,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">data =</span> .x))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<p>Make sure to <strong>understand</strong> the <code>map()</code> function here, it is <strong>completely central</strong> to functional programming with <code>purrr</code>:</p>
<ul>
<li>We need the <code>group_by()</code> to define which variable holds the elements to each of which we want to map</li>
<li><code>model_object</code> is a <strong>new</strong> variable, we are creating, which will contain the result of our call to the <code>map()</code> function</li>
<li><code>.x</code> to what <strong>existing</strong> (nested) variable are we mapping?</li>
<li><code>.f</code> which function do we want to map to each element in the <strong>existing</strong> (nested) variable?</li>
<li>Note that <code>log2_expr_level</code> and <code>early_metastasis</code> are variables “inside” the nested <code>data</code> variable</li>
</ul>
<p>Note, once again, this is conceptually a <em>super-tricky data structure</em>, not only do we have a <em>per gene</em> nested tibble, but now we also have a <em>per gene</em> nested model object - So please do make sure to discuss in your group, what is going on here, e.g.&nbsp;try running this and discuss what you see:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested <span class="sc">|&gt;</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"g2E09"</span>) <span class="sc">|&gt;</span> <span class="co"># Replace "g2E09" with whatever was YOUR favourite gene!</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(model_object)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidying-models" class="level4">
<h4 class="anchored" data-anchor-id="tidying-models">Tidying Models</h4>
<p>Excellent! Now we have a <em>per gene</em> model. Let us use the <code>broom</code> package to extract some information from each of the models. First, to get a better understanding of what is going on when calling the <code>tidy()</code> function, try running this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested <span class="sc">|&gt;</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Here, you should replace "g2E09" with whatever was YOUR favourite gene!</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"g2E09"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pull() on tibbles: This pulls out the model_object variable.</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   Note! This is a list, because we nested!</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(model_object) <span class="sc">|&gt;</span> </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pluck() on lists: From the list we got from the last step,</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   we "pluck" the first element</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The result of pluck, is a model object,</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   upon which we can call the tidy function</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tidy</span>(<span class="at">conf.int =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">conf.level =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we want to apply this <code>tidy()</code> function <em>per <code>model_object</code></em>:</p>
<ul>
<li><strong>T13: Scroll a bit back to where we created the <code>model_object</code> and see if you can translate that into mapping the <code>tidy()</code> function to the <code>model_object</code> variable, thereby creating a new variable <code>model_object_tidy</code> - This is tricky, so do make sure to discuss in your group how this can be done!</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Remember the parameters to the <code>tidy()</code> functions, which gives us the confidence intervals and defines corresponding level. Also, everything you need is in the previous layout of the <code>map()</code> function. Note, that the calls to the <code>pull()</code> and <code>pluck()</code> functions above, pertains to <strong>extracting</strong> from a tibble, which we do <strong>not</strong> want, on the contrary, we want <strong>all</strong> objects to be contained in our <code>gravier_clean_aug_long_nested</code> data</em>
</details>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,905 × 4
# Groups:   gene [2,905]
   gene   data               model_object model_object_tidy
   &lt;chr&gt;  &lt;list&gt;             &lt;list&gt;       &lt;list&gt;           
 1 g2E09  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 2 g7F07  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 3 g1A01  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 4 g3C09  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 5 g3H08  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 6 g1A08  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 7 g1B01  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 8 g1int1 &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
 9 g1E11  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
10 g8G02  &lt;tibble [168 × 2]&gt; &lt;lm&gt;         &lt;tibble [2 × 7]&gt; 
# ℹ 2,895 more rows</code></pre>
</div>
</div>
<p>Note, once again, this is conceptually a <em>super-tricky data structure</em>, not only do we have a <em>per gene</em> nested tibble, but now we also have a <em>per gene</em> nested model object and now also a nested tibble of <code>tidy</code>ed objects - So please again do make sure to discuss in your group, what is going on here, e.g.&nbsp;try running this and discuss what you see:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>gravier_clean_aug_long_nested <span class="sc">|&gt;</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">"g2E09"</span>) <span class="sc">|&gt;</span> <span class="co"># Replace "g2E09" with whatever was YOUR favourite gene!</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(model_object_tidy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="wrangling" class="level4">
<h4 class="anchored" data-anchor-id="wrangling">Wrangling</h4>
<p>We’re almost there - just a bit of wrangling to go!</p>
<p>Just as you saw that we could <code>nest()</code> on a variable (recall we did that for the <code>gene</code> variable), you can do the opposite and lo and behold, that is the <code>unnest()</code> function. Before we continue:</p>
<ul>
<li><strong>T14: Create a <code>dplyr</code> pipeline and save the result in a new variable called <code>gravier_estimates</code>: Use the <code>unnest()</code> function to unpack the <code>model_object_tidy</code></strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5,810 × 10
# Groups:   gene [2,905]
   gene  data     model_object term         estimate std.error statistic p.value
   &lt;chr&gt; &lt;list&gt;   &lt;list&gt;       &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
 1 g2E09 &lt;tibble&gt; &lt;lm&gt;         (Intercept)  -0.0162     0.0117    -1.38  1.69e-1
 2 g2E09 &lt;tibble&gt; &lt;lm&gt;         early_metas… -0.0343     0.0201    -1.71  8.99e-2
 3 g7F07 &lt;tibble&gt; &lt;lm&gt;         (Intercept)   0.0604     0.0139     4.35  2.36e-5
 4 g7F07 &lt;tibble&gt; &lt;lm&gt;         early_metas… -0.0185     0.0238    -0.778 4.38e-1
 5 g1A01 &lt;tibble&gt; &lt;lm&gt;         (Intercept)  -0.0290     0.0123    -2.35  1.99e-2
 6 g1A01 &lt;tibble&gt; &lt;lm&gt;         early_metas… -0.0367     0.0211    -1.73  8.47e-2
 7 g3C09 &lt;tibble&gt; &lt;lm&gt;         (Intercept)   0.0518     0.0145     3.58  4.55e-4
 8 g3C09 &lt;tibble&gt; &lt;lm&gt;         early_metas… -0.0148     0.0249    -0.595 5.53e-1
 9 g3H08 &lt;tibble&gt; &lt;lm&gt;         (Intercept)   0.0142     0.0128     1.11  2.69e-1
10 g3H08 &lt;tibble&gt; &lt;lm&gt;         early_metas…  0.00247    0.0220     0.112 9.11e-1
# ℹ 5,800 more rows
# ℹ 2 more variables: conf.low &lt;dbl&gt;, conf.high &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li><strong>T15: The again, create a <code>dplyr</code> pipeline and save the result in a the same <code>gravier_estimates</code> variable: Subset the rows to only get the slope term and then choose variables as displayed below, finally end with un-grouping your data, as we no longer need the groups</strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,905 × 5
   gene   p.value estimate conf.low conf.high
   &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
 1 g2E09   0.0899 -0.0343   -0.0739   0.00539
 2 g7F07   0.438  -0.0185   -0.0655   0.0285 
 3 g1A01   0.0847 -0.0367   -0.0784   0.00508
 4 g3C09   0.553  -0.0148   -0.0639   0.0343 
 5 g3H08   0.911   0.00247  -0.0410   0.0459 
 6 g1A08   0.859  -0.00363  -0.0438   0.0366 
 7 g1B01   0.279  -0.0218   -0.0615   0.0178 
 8 g1int1  0.666  -0.0113   -0.0627   0.0402 
 9 g1E11   0.106  -0.0329   -0.0728   0.00703
10 g8G02   0.633  -0.0108   -0.0555   0.0338 
# ℹ 2,895 more rows</code></pre>
</div>
</div>
<ul>
<li><strong>T16: To your <code>gravier_estimates</code> dataset, add a variable <code>q.value</code>, which is the result of calling the <code>p.adjust()</code> function on your <code>p.value</code> variable and also add an indicator variable denoting if a given gene is significant or not</strong></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,905 × 7
   gene   p.value estimate conf.low conf.high q.value is_significant
   &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;         
 1 g2E09   0.0899 -0.0343   -0.0739   0.00539       1 no            
 2 g7F07   0.438  -0.0185   -0.0655   0.0285        1 no            
 3 g1A01   0.0847 -0.0367   -0.0784   0.00508       1 no            
 4 g3C09   0.553  -0.0148   -0.0639   0.0343        1 no            
 5 g3H08   0.911   0.00247  -0.0410   0.0459        1 no            
 6 g1A08   0.859  -0.00363  -0.0438   0.0366        1 no            
 7 g1B01   0.279  -0.0218   -0.0615   0.0178        1 no            
 8 g1int1  0.666  -0.0113   -0.0627   0.0402        1 no            
 9 g1E11   0.106  -0.0329   -0.0728   0.00703       1 no            
10 g8G02   0.633  -0.0108   -0.0555   0.0338        1 no            
# ℹ 2,895 more rows</code></pre>
</div>
</div>
<p><em>But… <code>q.value</code>??? What are you on about??? For a nice primer on how multiple testing correction works, please read this article by <span class="citation" data-cites="Noble2009">Noble (<a href="references.html#ref-Noble2009" role="doc-biblioref">2009</a>)</span>!</em></p>
</section>
<section id="recap" class="level4">
<h4 class="anchored" data-anchor-id="recap">Recap</h4>
<p>If you (understandably by now) have lost a bit of overview of what is going on, let’s just re-iterate.</p>
<ol type="1">
<li>We have the <code>gravier</code> dataset, with the log2-expression levels for 2,905 genes of 168 patients of whom 111 did <em>not</em> have early metastasis and 57 who did</li>
<li>We are interested in investigating what genes are significantly up-/down-regulated between the patients with- and without early metastasis</li>
<li>First, we retrieved the data from the data repository, cleaned and augmented it and saved it to disk</li>
<li>Then pivoted the data, so we could work <em>per gene</em> (The <code>gene</code> variable)</li>
<li>Next, we grouped <em>per gene</em> and nested the data (The <code>data</code> variable)</li>
<li>Then, we fitted a linear model to each gene (The <code>model_object</code> variable)</li>
<li>Next, we used the <code>broom</code> package to tidy the fitted model incl.&nbsp;getting confidence intervals (The <code>model_object_tidy</code> variable)</li>
<li>Lastly, we extracted the model parameters, corrected for multiple testing and added and indicator for significant findings</li>
</ol>
<p>Now, we actually have everything we need to answer:</p>
<ul>
<li><strong>What genes are significantly up- or down-regulated between the patients with/without early metastasis?</strong></li>
</ul>
<p>In the following, we will use a level of significance of <span class="math inline">\(\alpha = 0.05\)</span> to provide this answer.</p>
</section>
<section id="visualise" class="level4">
<h4 class="anchored" data-anchor-id="visualise">Visualise</h4>
<p>Right, let’s get to it!</p>
<ul>
<li><strong>T17: Re-create this forest plot to finally reveal the results of your analysis</strong> <span style="color: red;">GROUP ASSIGNMENT</span> (Important, see: <a href="./assignments.html">how to</a>)</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab06_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Here, you will have to use your indicator variable to identify significant genes <strong>before</strong> plotting and then it would probably be prudent to take a look at the <code>fct_reorder()</code> function and <code>geom_errorbarh()</code> representation</em>
</details>
<ul>
<li><strong>T18: Re-create this volcano plot to finally reveal the results of your analysis</strong> <span style="color: red;">GROUP ASSIGNMENT part II</span></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="lab06_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Here,<strong>before</strong> plotting you will have to create a new label-variable which takes value <code>gene</code> for significant genes and otherwise is simply an empty string, which we denote by <code>""</code>. Also, perhaps the <code>ggrepel</code> package would be relevant for somehow adding text/labels</em>
</details>
</section>
</section>
</section>
<section id="group-assignment" class="level2">
<h2 class="anchored" data-anchor-id="group-assignment"><span style="color: red;">GROUP ASSIGNMENT</span></h2>
<p>For the group assignment this time, you will use <strong>T17</strong> and <strong>T18</strong> to again create a reproducible micro-report and make sure to:</p>
<ul>
<li><a href="https://r4bds.github.io/assignments.html">Read the assignment instructions</a></li>
<li><a href="https://r4bds.github.io/code_styling.html">Read and follow the code styling guidelines</a></li>
</ul>
</section>
<section id="optional" class="level2">
<h2 class="anchored" data-anchor-id="optional">Optional</h2>
<p><em>The following is only if you can find the time! But, perhaps this would be something interesting to revisit during the project period</em></p>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<ul>
<li>Go and visit this blog post by <a href="https://clauswilke.com/blog/2020/09/07/pca-tidyverse-style/">Claus O. Wilke, Professor of Integrative Biology</a>.</li>
<li>Your task is to work through the blog post using the <code>gravier</code> dataset to create a PCA analysis</li>
</ul>
</section>
<section id="k-means" class="level3">
<h3 class="anchored" data-anchor-id="k-means">K-means</h3>
<ul>
<li>Go and visit this <a href="https://www.tidymodels.org/learn/statistics/k-means/">K-means clustering with tidy data principles</a> post</li>
<li>Your task is to work through the blog post using the <code>gravier</code> dataset to create a K-means analysis</li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-Gravier2010" class="csl-entry" role="listitem">
Gravier, Eléonore, Gaëlle Pierron, Anne Vincent‐Salomon, Nadège Gruel, Virginie Raynal, Alexia Savignoni, Yann De Rycke, et al. 2010. <span>“A Prognostic DNA Signature for T1T2 Node‐negative Breast Cancer Patients.”</span> <em>Genes, Chromosomes and Cancer</em> 49 (12): 1125–34. <a href="https://doi.org/10.1002/gcc.20820">https://doi.org/10.1002/gcc.20820</a>.
</div>
<div id="ref-Noble2009" class="csl-entry" role="listitem">
Noble, William S. 2009. <span>“How Does Multiple Testing Correction Work?”</span> <em>Nature Biotechnology</em> 27 (12): 1135–37. <a href="https://doi.org/10.1038/nbt1209-1135">https://doi.org/10.1038/nbt1209-1135</a>.
</div>
<div id="ref-Nolan2020" class="csl-entry" role="listitem">
Nolan, Sean, Marissa Vignali, Mark Klinger, Jennifer N. Dines, Ian M. Kaplan, Emily Svejnoha, Tracy Craft, et al. 2020. <span>“A Large-Scale Database of t-Cell Receptor Beta (TCRβ) Sequences and Binding Associations from Natural and Synthetic Exposure to SARS-CoV-2.”</span> August. <a href="https://doi.org/10.21203/rs.3.rs-51964/v1">https://doi.org/10.21203/rs.3.rs-51964/v1</a>.
</div>
<div id="ref-Schorling1997" class="csl-entry" role="listitem">
Schorling, John B., Julienne Roach, Marjorie Siegel, Natalie Baturka, Dawn E. Hunt, Thomas M. Guterbock, and Herbert L. Stewart. 1997. <span>“A Trial of Church-Based Smoking Cessation Interventions for Rural African Americans.”</span> <em>Preventive Medicine</em> 26 (1): 92–101. <a href="https://doi.org/10.1006/pmed.1996.9988">https://doi.org/10.1006/pmed.1996.9988</a>.
</div>
<div id="ref-Torondel2016" class="csl-entry" role="listitem">
Torondel, Belen, Jeroen H. J. Ensink, Ozan Gundogdu, Umer Zeeshan Ijaz, Julian Parkhill, Faraji Abdelahi, Viet-Anh Nguyen, et al. 2016. <span>“Assessment of the Influence of Intrinsic Environmental and Geographical Factors on the Bacterial Ecology of Pit Latrines.”</span> <em>Microbial Biotechnology</em> 9 (2): 209–23. <a href="https://doi.org/10.1111/1751-7915.12334">https://doi.org/10.1111/1751-7915.12334</a>.
</div>
<div id="ref-Willems1997" class="csl-entry" role="listitem">
Willems, James P., J Terry Sanders, Dawn E. Hunt, and John B. Schorling. 1997. <span>“Prevalence of Coronary Heart Disease Risk Factors Among Rural Blacks: A Community-Based Study.”</span> <em>Southern Medical Journal</em> 90 (8): 814–20. <a href="https://doi.org/10.1097/00007611-199708000-00008">https://doi.org/10.1097/00007611-199708000-00008</a>.
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./lab05.html" class="pagination-link" aria-label="Lab 5: Data Wrangling II">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Lab 5: Data Wrangling II</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./lab07.html" class="pagination-link" aria-label="Lab 7: Collaborative Bio Data Science using GitHub via RStudio">
        <span class="nav-page-text">Lab 7: Collaborative Bio Data Science using GitHub via RStudio</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>R for Bio Data Science</p>
</div>   
    <div class="nav-footer-center">
<p>All content © 2023 - 2025 Leon Eyrich Jessen. All rights reserved</p>
</div>
    <div class="nav-footer-right">
<p>Full course description available at <a href="https://kurser.dtu.dk/course/22100">DTU Course Base</a>.</p>
</div>
  </div>
</footer>




</body></html>