<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.37">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>R for Bio Data Science - Lab 5: Data Wrangling II</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./lab06.html" rel="next">
<link href="./lab04.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

<link href="site_libs/table1-1.0/table1_defaults.css" rel="stylesheet">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Lab 5: Data Wrangling II</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">R for Bio Data Science</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle sidebar-tool" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome to R for Bio Data Science</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prologue.html" class="sidebar-item-text sidebar-link">Prologue</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">Getting Started</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./lab00.html" class="sidebar-item-text sidebar-link">Course Labs</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab01.html" class="sidebar-item-text sidebar-link">Lab 1: Course Intro &amp; the Very Basics</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab02.html" class="sidebar-item-text sidebar-link">Lab 2: Data Visualisation I</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab03.html" class="sidebar-item-text sidebar-link">Lab 3: Data Visualisation II</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab04.html" class="sidebar-item-text sidebar-link">Lab 4: Data Wrangling I</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab05.html" class="sidebar-item-text sidebar-link active">Lab 5: Data Wrangling II</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab06.html" class="sidebar-item-text sidebar-link">Lab 6: Functional Programming Purrr</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab07.html" class="sidebar-item-text sidebar-link">Lab 7: Scripting in a Reproducible and Collaborative Framework using GitHub via RStudio</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab08.html" class="sidebar-item-text sidebar-link">Lab 8: Creating a Simple R-package</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab09.html" class="sidebar-item-text sidebar-link">Lab 9 Creating a Simple Shiny Application</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./lab10.html" class="sidebar-item-text sidebar-link">Lab 10 Project Startup &amp; Industry Talks</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./resources.html" class="sidebar-item-text sidebar-link">Resources</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paths_and_projects.html" class="sidebar-item-text sidebar-link">Paths &amp; Projects</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./variable_assignment_in_r.html" class="sidebar-item-text sidebar-link">Variable Assignment in <code>R</code></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./code_styling.html" class="sidebar-item-text sidebar-link">Code Styling</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./external_resources.html" class="sidebar-item-text sidebar-link">External Resources</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./course_elements.html" class="sidebar-item-text sidebar-link">Course Elements</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignments.html" class="sidebar-item-text sidebar-link">Assignments</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./project_faq.html" class="sidebar-item-text sidebar-link">Project FAQ</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exam.html" class="sidebar-item-text sidebar-link">Exam</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#packages" id="toc-packages" class="nav-link active" data-scroll-target="#packages">Package(s)</a></li>
  <li><a href="#schedule" id="toc-schedule" class="nav-link" data-scroll-target="#schedule">Schedule</a></li>
  <li><a href="#learning-materials" id="toc-learning-materials" class="nav-link" data-scroll-target="#learning-materials">Learning Materials</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#sec-exercises" id="toc-sec-exercises" class="nav-link" data-scroll-target="#sec-exercises">Exercises</a>
  <ul class="collapse">
  <li><a href="#prologue" id="toc-prologue" class="nav-link" data-scroll-target="#prologue">Prologue</a></li>
  </ul></li>
  <li><a href="#intro" id="toc-intro" class="nav-link" data-scroll-target="#intro">Intro</a></li>
  <li><a href="#sec-background" id="toc-sec-background" class="nav-link" data-scroll-target="#sec-background">Background</a></li>
  <li><a href="#your-task-today" id="toc-your-task-today" class="nav-link" data-scroll-target="#your-task-today">Your Task Today</a></li>
  <li><a href="#getting-started" id="toc-getting-started" class="nav-link" data-scroll-target="#getting-started">Getting Started</a></li>
  <li><a href="#creating-the-micro-report" id="toc-creating-the-micro-report" class="nav-link" data-scroll-target="#creating-the-micro-report">Creating the Micro-Report</a>
  <ul class="collapse">
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#aim" id="toc-aim" class="nav-link" data-scroll-target="#aim">Aim</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load Libraries</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load Data</a></li>
  <li><a href="#data-description" id="toc-data-description" class="nav-link" data-scroll-target="#data-description">Data Description</a></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">Analysis</a></li>
  </ul></li>
  <li><a href="#epilogue" id="toc-epilogue" class="nav-link" data-scroll-target="#epilogue">Epilogue</a></li>
  <li><a href="#todays-assignment" id="toc-todays-assignment" class="nav-link" data-scroll-target="#todays-assignment"><span style="color: red;">Today’s Assignment</span></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Lab 5: Data Wrangling II</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="packages" class="level2">
<h2 class="anchored" data-anchor-id="packages">Package(s)</h2>
<ul>
<li><a href="https://dplyr.tidyverse.org">dplyr</a></li>
<li><a href="https://stringr.tidyverse.org">stringr</a></li>
<li><a href="https://tidyr.tidyverse.org">tidyr</a></li>
<li><a href="https://forcats.tidyverse.org">forcats</a></li>
<li><a href="https://patchwork.data-imaginist.com/index.html">patchwork</a></li>
<li><a href="https://omarwagih.github.io/ggseqlogo/">ggseqlogo</a></li>
<li><a href="https://cran.r-project.org/web/packages/table1/vignettes/table1-examples.html">table1</a></li>
</ul>
</section>
<section id="schedule" class="level2">
<h2 class="anchored" data-anchor-id="schedule">Schedule</h2>
<ul>
<li>08.00 - 08.30: <a href="https://raw.githack.com/r4bds/r4bds.github.io/main/recap_lab05.html">Recap of Lab 4</a></li>
<li>08.30 - 08.35: <a href="https://raw.githack.com/r4bds/r4bds.github.io/main/lecture_lab05.html">Lecture</a></li>
<li>08.35 - 08.45: Break</li>
<li>08.45 - 12.00: <a href="#sec-exercises">Exercises</a></li>
</ul>
</section>
<section id="learning-materials" class="level2">
<h2 class="anchored" data-anchor-id="learning-materials">Learning Materials</h2>
<p><em>Please prepare the following materials</em></p>
<ul>
<li>R4DS2e book: <a href="https://r4ds.hadley.nz/data-tidy">Chapter 6: Data Tidying</a>, <a href="https://r4ds.hadley.nz/strings">Chapter 15: Strings</a>, <a href="https://r4ds.hadley.nz/factors">Chapter 17: Factors</a>, <a href="https://r4ds.hadley.nz/joins">Chapter 20: Joins</a></li>
<li>Video: <a href="https://youtu.be/1ELALQlO-yM?t=465">Tidy Data and tidyr</a> - NB! Start at 7:45 and please note: <code>gather()</code> is now <code>pivot_longer()</code> and <code>spread()</code> is now <code>pivot_wider()</code></li>
<li>Video: <a href="https://www.youtube.com/watch?v=AuBgYDCg1Cg">Working with Two Datasets: Binds, Set Operations, and Joins</a></li>
<li>Video: <a href="https://www.youtube.com/watch?v=oIu5jK8DeX8&amp;list=PLiC1doDIe9rDwsUhd3FtN1XGCV2ES1xZ2">stringr</a> (Playlist with 7 short videos)</li>
</ul>
</section>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<p><em>A student who has met the objectives of the session will be able to:</em></p>
<ul>
<li>Understand and apply the various <code>str_*()</code>-functions for string manipulation</li>
<li>Understand and apply the family of <code>*_join()</code>-functions for combining data sets</li>
<li>Understand and apply <code>pivot_wider()</code> and <code>pivot_longer()</code></li>
<li>Use factors in context with plotting categorical data using <code>ggplot</code></li>
</ul>
</section>
<section id="sec-exercises" class="level2">
<h2 class="anchored" data-anchor-id="sec-exercises">Exercises</h2>
<section id="prologue" class="level3">
<h3 class="anchored" data-anchor-id="prologue">Prologue</h3>
<p>Today will not be easy! But please try to remember Hadley’s word-of-advise:</p>
<ul>
<li><em>“The bad news is, whenever you’re learning a new tool, for a long time, you’re going to suck! It’s gonna be very frustrating! But the good news is that that is typical and something that happens to everyone and it’s only temporary! Unfortunately, there is no way to going from knowing nothing about the subject to knowing something about a subject and being an expert in it without going through a period of great frustration and much suckiness! Keep pushing through!”</em> - H. Wickham <a href="https://www.youtube.com/watch?v=8SGif63VW6E">(dplyr tutorial at useR 2014, 4:10 - 4:48)</a></li>
</ul>
<div class="cell">

</div>
</section>
</section>
<section id="intro" class="level2">
<h2 class="anchored" data-anchor-id="intro">Intro</h2>
<p><em>We are upping the game here, so expect to get stuck at some of the questions. Remember - Discuss with your group how to solve the task, revisit the materials you prepared for today and naturally, the TAs and I are happy to nudge you in the right direction. Finally, remember… Have fun!</em></p>
<p>Remember what you have worked on so far:</p>
<ul>
<li>RStudio</li>
<li>Quarto</li>
<li>ggplot</li>
<li>filter</li>
<li>arrange</li>
<li>select</li>
<li>mutate</li>
<li>group_by</li>
<li>summarise</li>
<li>The pipe and creating pipelines</li>
<li>stringr</li>
<li>joining data</li>
<li>pivotting data</li>
</ul>
<p><em>That’s quite a lot! Well done - You’ve come quite far already! Remember to think about the above tools in the following as we will synthesise your learnings so far into an analysis!</em></p>
</section>
<section id="sec-background" class="level2">
<h2 class="anchored" data-anchor-id="sec-background">Background</h2>
<p>In the early 20s, the world was hit by the coronavirus disease 2019 (COVID-19) pandemic. The pandemic was caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). In Denmark the virus first confirmed case was on 27 February 2020.</p>
<p>While initially very little was known about the SARS-CoV-2 virus, we did know the general pathology of vira. Briefly, the virus invades the cells and hijacks the intra-cellular machinery. Using the hijacked machinery, components for new virus particles are produced, eventually being packed into the viral envelope and released from the infected cell. Some of these components, viral proteins, is broken down into smaller fragments called peptides by the proteasome. These peptides are transported into the endoplasmatic reticulum by the Transporter Associated with antigen Processing (TAP) protein complex. Here, they are aided by chaperones bound to the Major Histocompatilibty Complex class I (MHCI) and then across the Golgi Aparatus they finally get displayed on the surface of the cells. Note, in humans, MHC is also called Human Leukocyte Antigen (HLA) and represents the most diverse genes. Each of us have a total of 6 HLA-alleles, 3 from the maternal and 3 from the paternal side. These are further divided into 3 classes HLA-A, HLA-B and HLA-C and the combination of these constitute the HLA-haplotype for an individual. Once the peptide is bound to the MHC Class I at the cell surface and exposed, the MHCI-peptide complex can be recognised by CD8+ Cytotoxic T-Lymphocytes (CTLs) via the T-cell Receptor (TCR). If a cell displays peptides of viral origin, the CTL gets activated and via a cascade induces apoptosis (programmed cell death) of the infected cell. The proces is summarised in the figure below.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/mhc_class_I_antigen_presentation_pathway.png" class="img-fluid figure-img" style="width:60.0%"></p>
</figure>
</div>
<p><em>Image source: <a href="https://pubmed.ncbi.nlm.nih.gov/25688236/">10.3389/fmicb.2015.00021</a></em></p>
<p>The data we will be working with today contains data on sequenced T-cell receptors, viral antigens, HLA-haplotypes and clinical meta data for a cohort:</p>
<ul>
<li><a href="https://www.researchsquare.com/article/rs-51964/v1">A large-scale database of T-cell receptor beta (TCR<span class="math inline">\(\beta\)</span>) sequences and binding associations from natural and synthetic exposure to SARS-CoV-2</a></li>
</ul>
</section>
<section id="your-task-today" class="level2">
<h2 class="anchored" data-anchor-id="your-task-today">Your Task Today</h2>
<p>Today, we will emulate the situation, where you are working as a Bioinformatician / Bio Data Scientist and you have been given the data and the task of answering these two burning questions:</p>
<ol type="1">
<li>What characterises the peptides binding to the HLAs?</li>
<li>What characterises T-cell Receptors binding to the pMHC-complexes?</li>
</ol>
<p><span style="color: red;">GROUP ASSIGNMENT: Today, your assignment will be to create a micro-report on these 2 questions!</span></p>
</section>
<section id="getting-started" class="level2">
<h2 class="anchored" data-anchor-id="getting-started">Getting Started</h2>
<ol type="1">
<li><a href="https://teaching.healthtech.dtu.dk/22100/rstudio.php">Click here</a> to go to the course RStudio cloud server and login</li>
<li>Make sure you are in your <code>r_for_bio_data_science</code>-project, you can verify this in the upper right corner</li>
<li>In the same place as your <code>r_for_bio_data_science.Rproj</code>-file and existing <code>data</code>-folder, create a new folder and name it <code>doc</code></li>
<li>Go to the aforementioned <a href="https://www.researchsquare.com/article/rs-51964/v1">manuscript</a>. Download the PDF and upload it to your new <code>doc</code>-folder</li>
<li>Open the PDF and find the link to the data</li>
<li>Go to the data site (<em>Note, you may have to create and account to download, shouldn’t take too long</em>) . Find and download the file <code>ImmuneCODE-MIRA-Release002.1.zip</code> (<em>CAREFUL, do not download the superseded files</em>)</li>
<li>Unpack the downloaded file</li>
<li>Find the files <code>peptide-detail-ci.csv</code> and <code>subject-metadata.csv</code> and compress to <code>.zip</code>-files</li>
<li>Upload the compressed <code>peptide-detail-ci.csv.zip</code>- and <code>subject-metadata.csv.zip</code>-files to your <code>data</code>-folder in your RStudio Cloud session</li>
<li>Finally, once again, create a new Quarto document for today’s exercises, containing the sections:
<ol type="1">
<li>Background</li>
<li>Aim</li>
<li>Load Libraries</li>
<li>Load Data</li>
<li>Data Description</li>
<li>Analysis</li>
</ol></li>
</ol>
</section>
<section id="creating-the-micro-report" class="level2">
<h2 class="anchored" data-anchor-id="creating-the-micro-report">Creating the Micro-Report</h2>
<section id="background" class="level3">
<h3 class="anchored" data-anchor-id="background">Background</h3>
<p><em>Feel free to copy paste the one stated in the <a href="#sec-background">background</a>-section above</em></p>
</section>
<section id="aim" class="level3">
<h3 class="anchored" data-anchor-id="aim">Aim</h3>
<p><em>State the aim of the micro-report, i.e.&nbsp;what are the questions you are addressing?</em></p>
</section>
<section id="load-libraries" class="level3">
<h3 class="anchored" data-anchor-id="load-libraries">Load Libraries</h3>
<div class="cell">

</div>
<p><em>Load the libraries needed</em></p>
</section>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load Data</h3>
<p>Read the two data sets into variables <code>peptide_data</code> and <code>meta_data</code>.</p>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Think about which Tidyverse package deals with reading data and what are the file types we want to read here?</em>
</details>
<div class="cell">

</div>
</section>
<section id="data-description" class="level3">
<h3 class="anchored" data-anchor-id="data-description">Data Description</h3>
<p><em>It is customary to include a description of the data, helping the reader if the report, i.e.&nbsp;your stakeholder, to get an easy overview</em></p>
<section id="the-subject-meta-data" class="level4">
<h4 class="anchored" data-anchor-id="the-subject-meta-data">The Subject Meta Data</h4>
<p>Let’s take a look at the meta data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 30
   Experiment Subject `Cell Type` `Target Type` Cohort          Age Gender Race 
   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
 1 eMR18       360644 PBMC        C19_cI        COVID-19-B-N…    52 F      White
 2 eHO132          26 PBMC        C19_cI        COVID-19-Con…    65 F      White
 3 eQD129        2283 PBMC        C19_cI        COVID-19-Con…    60 F      White
 4 eHO140        1809 PBMC        C19_cI        COVID-19-Con…    NA &lt;NA&gt;   &lt;NA&gt; 
 5 eJL143       15845 naive_CD8   minigene_Set1 Healthy (No …    36 F      White
 6 eJL158         466 PBMC        C19_cI        COVID-19-Acu…    33 M      White
 7 eHO124        3819 PBMC        C19_cI        Healthy (No …    62 M      &lt;NA&gt; 
 8 eQD124         361 PBMC        C19_cI        COVID-19-B-N…    40 F      White
 9 eQD138         242 PBMC        C19_cII       COVID-19-Con…    NA &lt;NA&gt;   &lt;NA&gt; 
10 eAV100        1995 PBMC        C19_cII       COVID-19-Con…    29 F      &lt;NA&gt; 
# ℹ 22 more variables: `HLA-A...9` &lt;chr&gt;, `HLA-A...10` &lt;chr&gt;,
#   `HLA-B...11` &lt;chr&gt;, `HLA-B...12` &lt;chr&gt;, `HLA-C...13` &lt;chr&gt;,
#   `HLA-C...14` &lt;chr&gt;, DPA1...15 &lt;chr&gt;, DPA1...16 &lt;chr&gt;, DPB1...17 &lt;chr&gt;,
#   DPB1...18 &lt;chr&gt;, DQA1...19 &lt;chr&gt;, DQA1...20 &lt;chr&gt;, DQB1...21 &lt;chr&gt;,
#   DQB1...22 &lt;chr&gt;, DRB1...23 &lt;chr&gt;, DRB1...24 &lt;chr&gt;, DRB3...25 &lt;chr&gt;,
#   DRB3...26 &lt;chr&gt;, DRB4...27 &lt;chr&gt;, DRB4...28 &lt;chr&gt;, DRB5...29 &lt;chr&gt;,
#   DRB5...30 &lt;chr&gt;</code></pre>
</div>
</div>
<ul>
<li><p><strong>Q1: How many observations of how many variables are in the data?</strong></p></li>
<li><p><strong>Q2: Are there groupings in the variables, i.e.&nbsp;do certain variables “go together” somehow?</strong></p></li>
<li><p><strong>T1: Re-create this plot</strong></p></li>
</ul>
<p>Read this first:</p>
<ul>
<li><em>Think about: What is on the x-axis? What is on the y-axis? And also, it looks like we need to do some counting. Recall, that we can stick together a <code>dplyr</code>-pipeline with a call to <code>ggplot</code>, so here we will have to <code>count</code> of <code>Cohort</code> and <code>Gender</code> before plotting</em></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="lab05_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Does your plot look different somehow? Consider peeking at the hint…</p>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Perhaps not everyone agrees on how to denote <code>NA</code>s in data. I have seen <code>-99</code>, <code>-11</code>, <code>_</code> and so on… Perhaps this can be dealt with in the instance we read the data from the file? I.e. in the actual function call to your <code>read</code>-function. Recall, how can we get information on the parameters of a <code>?function</code></em>
</details>
<ul>
<li><strong>T2: Re-create this plot</strong></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="lab05_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Perhaps there is a function, which can <code>cut</code> continuous observations into a set of bins?</em>
</details>
<section id="stop-make-sure-you-handled-how-nas-are-denoted-in-the-data-before-proceeding-see-hint-below-t1" class="level5">
<h5 class="anchored" data-anchor-id="stop-make-sure-you-handled-how-nas-are-denoted-in-the-data-before-proceeding-see-hint-below-t1"><strong>STOP! Make sure you handled how <code>NA</code>s are denoted in the data before proceeding, see hint below T1</strong></h5>
<ul>
<li><strong>T3: Look at the data and create yet another plot as you see fit. Also skip the redundant variables <code>Subject</code>, <code>Cell Type</code> and <code>Target Type</code></strong></li>
</ul>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 27
   Experiment Cohort      Age Gender Race  `HLA-A...9` `HLA-A...10` `HLA-B...11`
   &lt;chr&gt;      &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;       
 1 eXL36      Healthy …    37 F      White A*01:01     A*02:01      B*15:01     
 2 eDH96      Healthy …    53 M      White A*02:01     A*29:01      B*15:01     
 3 eOX54      Healthy …    39 F      Afri… A*02:01     A*23:17      B*15:03     
 4 eOX46      Healthy …    45 M      White A*02:01     A*02:01      B*35:03     
 5 ePD73      Healthy …    37 F      White A*01:01     A*02:01      B*15:01     
 6 eEE243     Healthy …    32 F      &lt;NA&gt;  A*02:01     A*02:01      B*27:05     
 7 eLH42      COVID-19…    63 M      &lt;NA&gt;  A*02:01:01  A*23:01:01   B*07:02:01  
 8 eHO134     COVID-19…    36 M      White A*01:01:01  A*24:02:01   B*49:01:01  
 9 eQD117     COVID-19…    70 F      &lt;NA&gt;  A*02:01:01  A*24:02:01   B*35:02:01  
10 eEE217     Healthy …    32 F      White A*02:01     A*02:01      B*15:01     
# ℹ 19 more variables: `HLA-B...12` &lt;chr&gt;, `HLA-C...13` &lt;chr&gt;,
#   `HLA-C...14` &lt;chr&gt;, DPA1...15 &lt;chr&gt;, DPA1...16 &lt;chr&gt;, DPB1...17 &lt;chr&gt;,
#   DPB1...18 &lt;chr&gt;, DQA1...19 &lt;chr&gt;, DQA1...20 &lt;chr&gt;, DQB1...21 &lt;chr&gt;,
#   DQB1...22 &lt;chr&gt;, DRB1...23 &lt;chr&gt;, DRB1...24 &lt;chr&gt;, DRB3...25 &lt;chr&gt;,
#   DRB3...26 &lt;chr&gt;, DRB4...27 &lt;chr&gt;, DRB4...28 &lt;chr&gt;, DRB5...29 &lt;chr&gt;,
#   DRB5...30 &lt;chr&gt;</code></pre>
</div>
</div>
<p>Now, a classic way of describing a cohort, i.e.&nbsp;the group of subjects used for the study, is the so-called <code>table1</code> and while we could build this ourselves, this one time, in the interest of exercise focus and time, we are going to “cheat” and use an R-package, like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"table1"</span>) <span class="co"># &lt;= Yes, this should normally go at the beginning!</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Gender =</span> <span class="fu">factor</span>(Gender),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">Cohort =</span> <span class="fu">factor</span>(Cohort)) <span class="sc">|&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">table1</span>(<span class="at">x =</span> <span class="fu">formula</span>(<span class="sc">~</span> Gender <span class="sc">+</span> Age <span class="sc">+</span> Race <span class="sc">|</span> Cohort),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> _)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">

<div class="Rtable1"><table class="Rtable1">
<thead>
<tr>
<th class="rowlabel firstrow lastrow"></th>
<th class="firstrow lastrow"><span class="stratlabel">COVID-19-Acute<br><span class="stratn">(N=4)</span></span></th>
<th class="firstrow lastrow"><span class="stratlabel">COVID-19-B-Non-Acute<br><span class="stratn">(N=8)</span></span></th>
<th class="firstrow lastrow"><span class="stratlabel">COVID-19-Convalescent<br><span class="stratn">(N=90)</span></span></th>
<th class="firstrow lastrow"><span class="stratlabel">COVID-19-Exposed<br><span class="stratn">(N=3)</span></span></th>
<th class="firstrow lastrow"><span class="stratlabel">Healthy (No known exposure)<br><span class="stratn">(N=39)</span></span></th>
<th class="firstrow lastrow"><span class="stratlabel">Overall<br><span class="stratn">(N=144)</span></span></th>
</tr>
</thead>
<tbody>
<tr>
<td class="rowlabel firstrow">Gender</td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
</tr>
<tr>
<td class="rowlabel">F</td>
<td>1 (25.0%)</td>
<td>4 (50.0%)</td>
<td>33 (36.7%)</td>
<td>1 (33.3%)</td>
<td>17 (43.6%)</td>
<td>56 (38.9%)</td>
</tr>
<tr>
<td class="rowlabel">M</td>
<td>2 (50.0%)</td>
<td>3 (37.5%)</td>
<td>36 (40.0%)</td>
<td>0 (0%)</td>
<td>21 (53.8%)</td>
<td>62 (43.1%)</td>
</tr>
<tr>
<td class="rowlabel lastrow">Missing</td>
<td class="lastrow">1 (25.0%)</td>
<td class="lastrow">1 (12.5%)</td>
<td class="lastrow">21 (23.3%)</td>
<td class="lastrow">2 (66.7%)</td>
<td class="lastrow">1 (2.6%)</td>
<td class="lastrow">26 (18.1%)</td>
</tr>
<tr>
<td class="rowlabel firstrow">Age</td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
</tr>
<tr>
<td class="rowlabel">Mean (SD)</td>
<td>50.7 (17.0)</td>
<td>43.7 (7.74)</td>
<td>51.5 (15.3)</td>
<td>35.0 (NA)</td>
<td>33.3 (9.93)</td>
<td>44.9 (15.7)</td>
</tr>
<tr>
<td class="rowlabel">Median [Min, Max]</td>
<td>52.0 [33.0, 67.0]</td>
<td>42.0 [33.0, 53.0]</td>
<td>53.0 [21.0, 79.0]</td>
<td>35.0 [35.0, 35.0]</td>
<td>31.0 [21.0, 62.0]</td>
<td>42.0 [21.0, 79.0]</td>
</tr>
<tr>
<td class="rowlabel lastrow">Missing</td>
<td class="lastrow">1 (25.0%)</td>
<td class="lastrow">1 (12.5%)</td>
<td class="lastrow">21 (23.3%)</td>
<td class="lastrow">2 (66.7%)</td>
<td class="lastrow">0 (0%)</td>
<td class="lastrow">25 (17.4%)</td>
</tr>
<tr>
<td class="rowlabel firstrow">Race</td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
<td class="firstrow"></td>
</tr>
<tr>
<td class="rowlabel">African American</td>
<td>1 (25.0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>1 (2.6%)</td>
<td>2 (1.4%)</td>
</tr>
<tr>
<td class="rowlabel">White</td>
<td>2 (50.0%)</td>
<td>7 (87.5%)</td>
<td>13 (14.4%)</td>
<td>0 (0%)</td>
<td>28 (71.8%)</td>
<td>50 (34.7%)</td>
</tr>
<tr>
<td class="rowlabel">Asian</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>3 (3.3%)</td>
<td>0 (0%)</td>
<td>2 (5.1%)</td>
<td>5 (3.5%)</td>
</tr>
<tr>
<td class="rowlabel">Hispanic or Latino/a</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>1 (1.1%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>1 (0.7%)</td>
</tr>
<tr>
<td class="rowlabel">Native Hawaiian or Other Pacific Islander</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>1 (33.3%)</td>
<td>0 (0%)</td>
<td>1 (0.7%)</td>
</tr>
<tr>
<td class="rowlabel">Black or African American</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>3 (7.7%)</td>
<td>3 (2.1%)</td>
</tr>
<tr>
<td class="rowlabel">Mixed Race</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>0 (0%)</td>
<td>1 (2.6%)</td>
<td>1 (0.7%)</td>
</tr>
<tr>
<td class="rowlabel lastrow">Missing</td>
<td class="lastrow">1 (25.0%)</td>
<td class="lastrow">1 (12.5%)</td>
<td class="lastrow">73 (81.1%)</td>
<td class="lastrow">2 (66.7%)</td>
<td class="lastrow">4 (10.3%)</td>
<td class="lastrow">81 (56.3%)</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p><em>Note: This may look a bit odd initially, but if you render your document, you should be all good! Also, note how good this looks! If you have ever done a “Table 1” before, you know how painful they can be and especially if something changes in your cohort - Dynamic reporting to the rescue!</em></p>
<p>Lastly, before we proceed, the <code>meta_data</code> contains HLA data for both class I and class II (see background), but here we are <strong>only</strong> interested in class I, recall these are denoted <code>HLA-A</code>, <code>HLA-B</code> and <code>HLA-C</code>, so make sure to remove any non-class I, i.e.&nbsp;the one after, denoted <code>D</code>-something.</p>
<ul>
<li><strong>T4: Create a new version of the <code>meta_data</code>, which with respect to allele-data only contains information on class I and also fix the odd naming, e.g.&nbsp;<code>HLA-A...9</code> becomes <code>A1</code> oand <code>HLA-A...10</code> becomes <code>A2</code> and so on for <code>B1</code>, <code>B2</code>, <code>C1</code> and <code>C2</code>. Remember to assign this new data to the same <code>meta_data</code>-variable</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Which <code>tidyverse</code> function subsets variables? Perhaps there is a function, which somehow <code>matches</code> a set of variables? And perhaps for the initiated this is compatible with regular expressions (If you don’t know what this means - No worries! If you do, see if you utilise this to simplify your variable selection)</em>
</details>
<div class="cell">

</div>
<p>Before we proceed, this is the data we will carry on with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 11
   Experiment Cohort        Age Gender Race  A1    A2    B1    B2    C1    C2   
   &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
 1 eMR25      COVID-19-C…    21 F      &lt;NA&gt;  ""    ""    ""    ""    ""    ""   
 2 eHO125     COVID-19-C…    52 M      &lt;NA&gt;  "A*0… "A*0… "B*3… "B*4… "C*0… "C*0…
 3 eXL31      Healthy (N…    28 M      White "A*0… "A*2… "B*0… "B*4… "C*0… "C*1…
 4 eMR14      COVID-19-C…    NA &lt;NA&gt;   &lt;NA&gt;  "A*0… "A*2… "B*0… "B*5… "C*0… "C*0…
 5 eHO134     COVID-19-C…    36 M      White "A*0… "A*2… "B*4… "B*5… "C*0… "C*0…
 6 eHO127     COVID-19-C…    28 M      &lt;NA&gt;  "A*2… "A*2… "B*4… "B*5… "C*0… "C*1…
 7 eJL148     COVID-19-C…    41 F      &lt;NA&gt;  "A*0… "A*0… "B*0… "B*1… "C*0… "C*0…
 8 ePD80      COVID-19-C…    67 M      &lt;NA&gt;  "A*0… "A*6… "B*1… "B*4… "C*0… "C*1…
 9 eQD118     COVID-19-C…    67 F      &lt;NA&gt;  "A*0… "A*1… "B*4… "B*5… "C*0… "C*1…
10 eOX49      Healthy (N…    21 M      White "A*0… "A*2… "B*4… "B*5… "C*0… "C*0…</code></pre>
</div>
</div>
<p>Now, we have a beautiful <code>tidy</code>-dataset, recall that this entails, that each row is an observation, each column is a variable and each cell holds one value.</p>
</section>
</section>
<section id="the-peptide-details-data" class="level4">
<h4 class="anchored" data-anchor-id="the-peptide-details-data">The Peptide Details Data</h4>
<p>Let’s start with simply having a look see:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   `TCR BioIdentity`            TCR Nucleotide Seque…¹ Experiment `ORF Coverage`
   &lt;chr&gt;                        &lt;chr&gt;                  &lt;chr&gt;      &lt;chr&gt;         
 1 CASSQEVLANSHEQYF+TCRBV04-03… ACCCTGCAGCCAGAAGACTCG… eXL31      ORF7b         
 2 CASSGGLSYEQYF+TCRBV25-01+TC… ACCCTGGAGTCTGCCAGGCCC… ePD83      membrane glyc…
 3 CATSDSRPDVNTEAFF+TCRBV24-01… TCTGCCATCCCCAACCAGACA… eEE224     ORF1ab        
 4 CASSLAGVTEAFF+TCRBV07-06+TC… ACGATCCAGCGCACAGAGCAG… eEE240     membrane glyc…
 5 CASSLAGNNEQFF+TCRBV27-01+TC… ATCCTGGAGTCGCCCAGCCCC… eXL31      ORF1ab        
 6 CASSSGSPTGYEQYF+TCRBV06-05+… NTGTCGGCTGCTCCCTCCCAG… eOX43      ORF1ab        
 7 CASSSLPGGGEQYF+TCRBV13-01+T… ATGAGCTCCTTGGAGCTGGGG… eHH175     ORF1ab        
 8 CASSLSAGNTIYF+TCRBV27-01+TC… ATCCTGGAGTCGCCCAGCCCC… eXL30      surface glyco…
 9 CSGTSSTDTQYF+TCRBV20-X+TCRB… CTGACAGTGACCAGTGCCCAT… eXL30      ORF7b         
10 CASGTQNTGELFF+TCRBV05-01+TC… AATGTGAGCACCTTGGAGCTG… eHO141     surface glyco…
# ℹ abbreviated name: ¹​`TCR Nucleotide Sequence`
# ℹ 3 more variables: `Amino Acids` &lt;chr&gt;, `Start Index in Genome` &lt;dbl&gt;,
#   `End Index in Genome` &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li><strong>Q3: How many observations of how many variables are in the data?</strong></li>
</ul>
<p>This is a rather big data set, so let us start with two “tricks” to handle this, first:</p>
<ol type="1">
<li><em>Write the data back into your <code>data</code>-folder, using the filename <code>peptide-detail-ci.csv.gz</code>, note the appending of <code>.gz</code>, which is automatically recognised and results in gz-compression</em></li>
<li><em>Now, check in your data folder, that you have two files <code>peptide-detail-ci.csv</code> and <code>peptide-detail-ci.csv.gz</code>, delete the former</em></li>
<li><em>Adjust your reading-the-data-code in the “Load Data”-section, to now read in the <code>peptide-detail-ci.csv.gz</code>-file</em></li>
</ol>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Just as you can <code>read</code> a file, you can of course also <code>write</code> a file. Note the filetype we want to write here is <code>csv</code>. If you in the console type e.g.&nbsp;<code>readr::wr</code> and then hit the <code>tab</code>-button, you will see the different functions for writing different filetypes</em>
</details>
<p>Then:</p>
<ul>
<li><strong>T5: As before, let’s immediately subset the <code>peptide_data</code> to the variables of interest: <code>TCR BioIdentity</code>, <code>Experiment</code> and <code>Amino Acids</code>. Remember to assign this new data to the same <code>peptide_data</code>-variable to avoid cluttering your environment with redundant variables. Bonus: Did you know you can click the <code>Environment</code> pane and see which variables you have?</strong></li>
</ul>
<div class="cell">

</div>
<p>Once again, before we proceed, this is the data we will carry on with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   Experiment `TCR BioIdentity`                           `Amino Acids`         
   &lt;chr&gt;      &lt;chr&gt;                                       &lt;chr&gt;                 
 1 eLH46      CASSPDVEPTDTQYF+TCRBV06-05+TCRBJ02-03       FAYANRNRF,LQFAYANRNR,…
 2 eEE228     CASSPSSGSPYEQYF+TCRBV18-01+TCRBJ02-07       FPNITNLCPF,QPTESIVRF,…
 3 eEE226     CASSHYLDTLRGYTF+TCRBV03-01/03-02+TCRBJ01-02 KLSYGIATV             
 4 eEE226     CASSWGLGYEQYF+TCRBV05-05+TCRBJ02-07         LLTDEMIAQY,LTDEMIAQY,…
 5 eLH48      CASSIGSTDTQYF+TCRBV07-09+TCRBJ02-03         AYKTFPPTEPK,KTFPPTEPK 
 6 eOX43      CASTMGLAVYNEQFF+TCRBV03-01/03-02+TCRBJ02-01 FLPRVFSAV             
 7 eXL31      CASSLSGSNEQFF+TCRBV27-01+TCRBJ02-01         QLMCQPILL,QLMCQPILLL  
 8 eMR17      CASSPDSSAASTDTQYF+TCRBV27-01+TCRBJ02-03     HTTDPSFLGRY           
 9 eOX54      CASRQGPNSYEQYF+TCRBV07-08+TCRBJ02-07        ELYSPIFLI,LYSPIFLIV,Q…
10 eOX49      CASSPYSNQPQHF+TCRBV06-X+TCRBJ01-05          AFLLFLVLI,FLAFLLFLV,F…</code></pre>
</div>
</div>
<ul>
<li><p><strong>Q4: Is this tidy data? Why/why not?</strong></p></li>
<li><p><strong>T6: See if you can find a way to create the below data, from the above</strong></p></li>
</ul>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="at">size =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
   Experiment CDR3b             V_gene     J_gene     `Amino Acids`             
   &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;                     
 1 eEE240     CASSSRRDRLYNEQFF  TCRBV13-01 TCRBJ02-01 MVMCGGSLYV,VMCGGSLYV      
 2 eXL30      CATSDRTGAYEQYF    TCRBV24-01 TCRBJ02-07 AFLLFLVLI,FLAFLLFLV,FYLCF…
 3 eHO136     CASSSTSGSLNEQYF   TCRBV07-06 TCRBJ02-07 LLTDEMIAQY,LTDEMIAQY,LTDE…
 4 eEE226     CSATPTPGHYPYEQYF  TCRBV20-X  TCRBJ02-07 APKEIIFL,KEIIFLEGETL      
 5 eEE228     CASSHPTSGSLGELFF  TCRBV04-03 TCRBJ02-02 FVCNLLLLFV,LLFVTVYSHL,TVY…
 6 eOX46      CASSSGMRNTEAFF    TCRBV12-X  TCRBJ01-01 AFLLFLVLI,FLAFLLFLV,FYLCF…
 7 eQD124     CAWSAGYEQYF       TCRBV30-01 TCRBJ02-07 HTTDPSFLGRY               
 8 eOX54      CASSLAGGSYEQYF    TCRBV05-01 TCRBJ02-07 VTIAEILLI                 
 9 eXL37      CSARGATRGGAGETQYF TCRBV20-X  TCRBJ02-05 AFLLFLVLI,FLAFLLFLV,FYLCF…
10 eJL148     CASTKSSGAYNEQFF   TCRBV19-01 TCRBJ02-01 LSPRWYFYY,SPRWYFYYL       </code></pre>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>First: Compare the two datasets and identify what happened? Did any variables “dissappear” and did any “appear”? Ok, so this is a bit tricky, but perhaps there is a function to <code>separate</code> a composit (untidy) <code>col</code>umn <code>into</code> a set of new variables based on a <code>sep</code>arator? But what is a <code>sep</code>arator? Just like when you read a file with <code>C</code>omma <code>S</code>eparated <code>V</code>alues, a separator denotes how a composite string is divided into fields. So look for such a repeated values, which seem to indeed separate such fields. Also, be aware, that character, which can mean more than one thing, may need to be “escaped” using an initial two backslashed, i.e.&nbsp;“\x”, where x denotes the character needing to be “escaped”</em>
</details>
<ul>
<li><strong>T7: Add a variable, which counts how many peptides are in each observation of <code>Amino Acids</code></strong></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>We have been working with the <code>stringr</code>-package, perhaps the contains a function to somehow count the number of occurrences of a given character in a string? Again, remember you can type e.g.&nbsp;<code>stringr::str_</code> and then hit the <code>tab</code>-button to see relevant functions</em>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="at">size =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 6
   Experiment CDR3b                 V_gene       J_gene `Amino Acids` n_peptides
   &lt;chr&gt;      &lt;chr&gt;                 &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;              &lt;dbl&gt;
 1 eEE226     CASSQVLSGPEQFF        TCRBV14-01   TCRBJ… APKEIIFL,KEI…          2
 2 eEE226     CASRSLLIGTGGPRMNTEAFF TCRBV06-X    TCRBJ… DFLEYHDVR,ED…          5
 3 eXL30      CASSWGLVYEQYF         TCRBV05-05   TCRBJ… LLTDEMIAQY,L…          4
 4 eOX54      CASSFSVEQFF           TCRBV12-03/… TCRBJ… AFPFTIYSL,GY…          7
 5 eMR18      CSAPDGYYGYTF          TCRBV20-X    TCRBJ… CELYHYQECV,I…          4
 6 eXL30      CASSLNHTQYF           TCRBV05-06   TCRBJ… YLNTLTLAV              1
 7 eAV93      CASRMGTSEYEQYF        TCRBV07-08   TCRBJ… CELYHYQECV,I…          4
 8 eMR23      CASSLAGARELFF         TCRBV05-01   TCRBJ… LSPRWYFYY,SP…          2
 9 eLH50      CACLEDYEQYF           TCRBV30-01   TCRBJ… RIRGGDGKM,RI…          2
10 eOX52      CSVEVGDGYEQYF         TCRBV29-01   TCRBJ… AFPFTIYSL,GY…          7</code></pre>
</div>
</div>
<ul>
<li><strong>T8: Re-create the following plot</strong></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="lab05_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<ul>
<li><p><strong>Q4: What is the maximum number of peptides assigned to one observation?</strong></p></li>
<li><p><strong>T9: Using the <code>str_c</code>- and the <code>seq</code>-functions, re-create the below</strong></p></li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "peptide_1" "peptide_2" "peptide_3" "peptide_4" "peptide_5"</code></pre>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>If you’re uncertain on how a function works, try going into the console and in this case e.g.&nbsp;type <code>str_c("a", "b")</code> and <code>seq(from = 1, to = 3)</code> and see if you combine these?</em>
</details>
<ul>
<li><strong>T10: Use, what you learned about separating in T6 and the vector-of-strings you created in T9 adjusted to the number from Q4 to create the below data</strong></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>In the console, write <code>?separate</code> and think about how you used it earlier. Perhaps you can not only specify a vector to separate <code>into</code>, but also specify a function, which returns a vector?</em>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="at">size =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 18
   Experiment CDR3b        V_gene J_gene peptide_1 peptide_2 peptide_3 peptide_4
   &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;    
 1 eHO130     CASSQALSSYN… TCRBV… TCRBJ… FVDGVPFVV &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;     
 2 eOX54      CASSLVPTGGN… TCRBV… TCRBJ… DFLEYHDVR EDFLEYHD… LEYHDVRVV LEYHDVRV…
 3 eMR13      CASSLSTPQWD… TCRBV… TCRBJ… HTTDPSFL… &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;     
 4 eEE240     CATSKTRRDNE… TCRBV… TCRBJ… FVDGVPFVV &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;     
 5 eEE224     CSGERETSYNE… TCRBV… TCRBJ… AFLLFLVLI FLAFLLFLV FYLCFLAFL FYLCFLAF…
 6 eMR12      CASSLSSGVVD… TCRBV… TCRBJ… HTTDPSFL… &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;     
 7 eDH105     CASSAGVVTEA… TCRBV… TCRBJ… ALALLLLDR GDAALALL… LALLLLDRL LLLDRLNQL
 8 eEE240     CASSYDSGTGL… TCRBV… TCRBJ… KLSYGIATV &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;     
 9 eMR18      unproductive TCRBV… TCRBJ… YLQPRTFL  YLQPRTFLL YYVGYLQP… &lt;NA&gt;     
10 eQD124     CASPTSGRVNE… TCRBV… TCRBJ… ATSRTLSYY ATSRTLSY… TSRTLSYYK TVATSRTL…
# ℹ 10 more variables: peptide_5 &lt;chr&gt;, peptide_6 &lt;chr&gt;, peptide_7 &lt;chr&gt;,
#   peptide_8 &lt;chr&gt;, peptide_9 &lt;chr&gt;, peptide_10 &lt;chr&gt;, peptide_11 &lt;chr&gt;,
#   peptide_12 &lt;chr&gt;, peptide_13 &lt;chr&gt;, n_peptides &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li><p><strong>Q5: Now, presumable you got a warning, discuss in your group why that is?</strong></p></li>
<li><p><strong>Q6: With respect to <code>peptide_n</code>, discuss in your group, if this is wide- or long-data?</strong></p></li>
</ul>
<p>Now, finally we will use the what we prepared for today, data-pivotting. There are two functions, namely <code>pivot_wider()</code> and <code>pivot_longer()</code>. Also, now, we will use a trick when developing ones data pipeline, while working with new functions, that on might not be completely comfortable with. You have seen the <code>sample_n()</code>-function several times above and we can use that to randomly sample <code>n</code>-observations from data. This we can utilise to work with a smaller data set in the development face and once we are ready, we can increase this <code>n</code> gradually to see if everything continues to work as anticipated.</p>
<ul>
<li><p><strong>T11: Using the <code>peptide_data</code>, run a few <code>sample_n()</code>-calls with varying degree of <code>n</code> to make sure, that you get a feeling for what is going on</strong></p></li>
<li><p><strong>T12: From the <code>peptide_data</code> data above, with peptide_1, peptide_2, etc. create this data set using one of the data-pivotting functions. Remember to start initially with sampling a smaller data set and then work on that first! Also, once you’re sure you’re good to go, reuse the <code>peptide_data</code>-variable as we don’t want huge redundant data sets floating around in our environment</strong></p></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<p>If the pivotting is not clear at all, then do what I do, create some example data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>my_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">str_c</span>(<span class="st">"id_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">var_1 =</span> <span class="fu">round</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="dv">1</span>),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">var_2 =</span> <span class="fu">round</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="dv">1</span>),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">var_3 =</span> <span class="fu">round</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>),<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>…and then play around with that. A small set like the one above is easy to handle, so perhaps start with that and then pivot back and forth a few times using the <code>pivot_wider()</code>-/<code>pivot_longer()</code>-functions. Use the <code>View()</code>-function to inspect and get a better overview of the results of pivotting.</p>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   Experiment CDR3b              V_gene      J_gene n_peptides peptide_n peptide
   &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;       &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;  
 1 eOX54      CASSSAIGTSGRPYEQYF TCRBV07-09  TCRBJ…          1 peptide_… &lt;NA&gt;   
 2 eEE224     CASSVAGTSTDTQYF    TCRBV09-01  TCRBJ…          5 peptide_6 &lt;NA&gt;   
 3 eEE226     CASSSPGQGFYEQYF    TCRBV28-01  TCRBJ…          6 peptide_1 DGVYFA…
 4 eXL31      CASTTGRGLNTGELFF   TCRBV03-01… TCRBJ…          4 peptide_3 LITLAT…
 5 ePD85      CASSTGGGLSYEQYF    TCRBV19-01  TCRBJ…          3 peptide_7 &lt;NA&gt;   
 6 eEE224     CSVEDGLGNIQYF      TCRBV29-01  TCRBJ…          4 peptide_8 &lt;NA&gt;   
 7 eXL27      CSAGGARVDEQFF      TCRBV20-X   TCRBJ…         11 peptide_… YLCFLA…
 8 eOX54      CAWKYSRTEAFF       TCRBV30-01  TCRBJ…          1 peptide_6 &lt;NA&gt;   
 9 eOX54      CSASVGDSSYEQYF     TCRBV20-01  TCRBJ…          1 peptide_… &lt;NA&gt;   
10 ePD100     CASSLSQGESGELFF    TCRBV05-01  TCRBJ…          4 peptide_6 &lt;NA&gt;   </code></pre>
</div>
</div>
<ul>
<li><p><strong>Q7: You will see some <code>NA</code>s in the <code>peptide</code>-variable, discuss in your group from where these arise?</strong></p></li>
<li><p><strong>Q8: How many rows and columns now and how does this compare with Q3? Discuss why/why not it is different?</strong></p></li>
<li><p><strong>T13: Now, loose the redundant variables <code>n_peptides</code> and <code>peptide_n</code> and also get rid of the <code>NA</code>s in the <code>peptide</code>-column and make sure, that we only have unique observations, i.e.&nbsp;there are no repeated rows/observations</strong></p></li>
</ul>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
   Experiment CDR3b              V_gene           J_gene     peptide   
   &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;            &lt;chr&gt;      &lt;chr&gt;     
 1 eAV91      CASSSGQGDGYTF      TCRBV09-01       TCRBJ01-02 VEHVTFFIY 
 2 eQD131     CASRVFGTDTDTQYF    TCRBV11-02       TCRBJ02-03 INVFAFPFTI
 3 eOX43      CASSLVAGSPAYEQYF   TCRBV27-01       TCRBJ02-07 AFLLFLVLI 
 4 eEE240     CASSQEGGGSGNEQFF   TCRBV27-01       TCRBJ02-01 SELVIGAVIL
 5 eEE240     CSASGQGYEQYF       TCRBV20-X        TCRBJ02-07 IELSLIDFYL
 6 eXL27      CASTATGAWGANVLTF   TCRBV19-01       TCRBJ02-06 LLFLAFVVFL
 7 eOX52      CAWEGVQAGGPRSYEQYF TCRBV30-01       TCRBJ02-07 FVDGVPFVV 
 8 eEE226     CASSQEGGGAFETQYF   TCRBV03-01/03-02 TCRBJ02-05 WPVTLACFVL
 9 eOX52      CASSLGGLNEQFF      TCRBV27-01       TCRBJ02-01 LYSPIFLIV 
10 eXL37      CSATGGGQPQHF       TCRBV20-X        TCRBJ01-05 IELSLIDFYL</code></pre>
</div>
</div>
<ul>
<li><strong>Q8: Now how many rows and columns and is this data tidy? Discuss in your group why/why not?</strong></li>
</ul>
<p>Again, we turn to the <code>stringr</code>-package, as we need to make sure that the sequence data does indeed only contain valid characters. There are a total of 20 proteogenic amino acids, which we symbolise using <code>ARNDCQEGHILKMFPSTWYV</code>.</p>
<ul>
<li><strong>T14: Use the <code>str_detect()</code>-function to <code>filter</code> the <code>CDR3b</code> and <code>peptide</code> variables using a <code>pattern</code> of <code>[^ARNDCQEGHILKMFPSTWYV]</code> and then play with the <code>negate</code>-parameter so see what happens</strong></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Again, try to play a bit around with the function in the console, type e.g.&nbsp;<code>str_detect(string = "ARND", pattern = "A")</code> and <code>str_detect(string = "ARND", pattern = "C")</code> and then recall, that the <code>filter</code>-function requires a logical vector, i.e.&nbsp;a vector of <code>TRUE</code> and <code>FALSE</code> to filter the rows</em>
</details>
<ul>
<li><strong>T15: Add two new variables to the data, <code>k_CDR3b</code> and <code>k_peptide</code> each signifying the length of the respective sequences</strong></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Again, we’re working with strings, so perhaps there is a package of interest and perhaps in that package, there is a function, which can get the length of a string?</em>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   Experiment CDR3b            V_gene           J_gene peptide k_CDR3b k_peptide
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;
 1 eOX52      CASSTTGEGYTF     TCRBV12-03/12-04 TCRBJ… INVFAF…      12        10
 2 eQD134     CASSLATDREAFF    TCRBV05-06       TCRBJ… ELIRQG…      13         9
 3 eOX49      CAISESVGIGQYF    TCRBV10-03       TCRBJ… LIVNSV…      13        10
 4 eOX49      CSVALNNEQFF      TCRBV29-01       TCRBJ… YLCFLA…      11         9
 5 eEE240     CASSQLAGGPNEQFF  TCRBV14-01       TCRBJ… FYLCFL…      15         9
 6 eOX49      CASSAGTPYEQYF    TCRBV06-02/06-03 TCRBJ… GYINVF…      13        10
 7 eXL31      CASSLSGELFF      TCRBV12-03/12-04 TCRBJ… AFPFTI…      11         9
 8 eXL27      CASSIDLYGNAYGYTF TCRBV19-01       TCRBJ… AFLLFL…      16         9
 9 eHO136     CASSSEQLNTEAFF   TCRBV27-01       TCRBJ… FYLCFL…      14         9
10 eOX46      CSALAGGNEQFF     TCRBV29-01       TCRBJ… TLACFV…      12        10</code></pre>
</div>
</div>
<ul>
<li><strong>T16: Re-create this plot</strong></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="lab05_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<ul>
<li><p><strong>Q9: What is the most predominant length of the CDR3b-sequences?</strong></p></li>
<li><p><strong>T17: Re-create this plot</strong></p></li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="lab05_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<ul>
<li><p><strong>Q10: What is the most predominant length of the peptide-sequences?</strong></p></li>
<li><p><strong>Q11: Discuss in your group, if this data set is tidy or not?</strong></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   Experiment CDR3b             V_gene          J_gene peptide k_CDR3b k_peptide
   &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;
 1 eEE228     CASSPSGGELFF      TCRBV12-03/12-… TCRBJ… MGYINV…      12         9
 2 eXL30      CSVGVEAKNIQYF     TCRBV29-01      TCRBJ… FVCNLL…      13        10
 3 eHO135     CATSVTFGAGYTGELFF TCRBV24-01      TCRBJ… SPRWYF…      17         9
 4 eXL31      CSATEAGDETQYF     TCRBV20-X       TCRBJ… YLCFLA…      13         9
 5 eEE226     CASSGRVQETQYF     TCRBV19-01      TCRBJ… APKEII…      13         8
 6 eOX49      CASSPQGQGQLFF     TCRBV04-01      TCRBJ… IYKTPP…      13        10
 7 eHO128     CASSLQSPSSAGNEQFF TCRBV27-01      TCRBJ… INFVRI…      17         9
 8 eOX43      CASRYTVNTEAFF     TCRBV27-01      TCRBJ… SELVIG…      13        10
 9 eEE240     CASSYTGDSFGELFF   TCRBV27-01      TCRBJ… SELVIG…      15        10
10 eXL27      CASSQDIGTGGNNEQFF TCRBV04-01      TCRBJ… ELYSPI…      17         9</code></pre>
</div>
</div>
</section>
<section id="creating-one-data-set-from-two-data-sets" class="level4">
<h4 class="anchored" data-anchor-id="creating-one-data-set-from-two-data-sets">Creating one data set from two data sets</h4>
<p>Before we move onto using the family of <code>*_join</code>-functions you prepared for today, we will just take a quick peek at the meta data again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 11
   Experiment Cohort        Age Gender Race  A1    A2    B1    B2    C1    C2   
   &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
 1 eQD139     COVID-19-C…    NA &lt;NA&gt;   &lt;NA&gt;  A*01… A*29… B*56… B*57… C*01… C*06…
 2 eQD127     COVID-19-C…    61 F      &lt;NA&gt;  A*02… A*03… B*27… B*35… C*02… C*04…
 3 eQD137     COVID-19-C…    NA &lt;NA&gt;   &lt;NA&gt;  A*01… A*02… B*37… B*44… C*06… C*16…
 4 eQD125     COVID-19-C…    44 M      &lt;NA&gt;  A*01… A*11… B*15… B*55… C*01… C*08…
 5 eXL37      Healthy (N…    33 M      White A*02… A*03… B*40… B*44… C*03… C*16…
 6 eJL157     COVID-19-C…    27 F      &lt;NA&gt;  A*11… A*11… B*07… B*18… C*07… C*07…
 7 eJL158     COVID-19-A…    33 M      White A*02… A*24… B*15… B*40… C*03… C*15…
 8 eJL151     COVID-19-C…    79 F      &lt;NA&gt;  A*24… A*68… B*15… B*40… C*03… C*03…
 9 eMR18      COVID-19-B…    52 F      White A*02… A*68… B*14… B*18… C*07… C*08…
10 eLH57      COVID-19-C…    NA &lt;NA&gt;   &lt;NA&gt;  A*01… A*03… B*07… B*37… C*06… C*07…</code></pre>
</div>
</div>
<p>Remember you can scroll in the data.</p>
<ul>
<li><strong>Q12: Discuss in your group, if this data with respect to the <code>A1</code>-, <code>A2</code>-, <code>B1</code>-, <code>B2</code>-, <code>C1-</code> and <code>C2</code>-variables is a wide- or a long-data format?</strong></li>
</ul>
<p>As with the <code>peptide_data</code>, we will now have to use data-pivotting again. I.e.:</p>
<ul>
<li><strong>T18: use either the <code>pivot_wider</code>- or <code>pivot_longer</code>-function to create the following data:</strong></li>
</ul>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   Experiment Cohort                        Age Gender Race         Gene  Allele
   &lt;chr&gt;      &lt;chr&gt;                       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;        &lt;chr&gt; &lt;chr&gt; 
 1 eQD129     COVID-19-Convalescent          60 F      White        B1    B*44:…
 2 eDH105     COVID-19-Convalescent          32 F      &lt;NA&gt;         C2    C*08:…
 3 ePD87      COVID-19-Convalescent          47 M      White        A1    A*03:…
 4 ePD76      Healthy (No known exposure)    33 M      White        C2    C*03:…
 5 ePD86      COVID-19-Convalescent          58 M      White        B1    B*44:…
 6 eOX49      Healthy (No known exposure)    21 M      White        C1    C*03:…
 7 ePD86      COVID-19-Convalescent          58 M      White        C2    C*14:…
 8 eQD127     COVID-19-Convalescent          61 F      &lt;NA&gt;         C2    C*04:…
 9 eHH170     Healthy (No known exposure)    24 F      Black or Af… C1    C*04:…
10 eHO134     COVID-19-Convalescent          36 M      White        C2    C*07:…</code></pre>
</div>
</div>
<p>Remember, what we are aiming for here, is to create one data set from two. So:</p>
<ul>
<li><strong>Q13: Discuss in your group, which variable(s?) define the same observations between the <code>peptide_data</code> and the <code>meta_data</code>?</strong></li>
</ul>
<p>Once you have agreed upon <code>Experiment</code>, then use that knowledge to subset the <code>meta_data</code> to the variables-of-interest:</p>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 2
   Experiment Allele    
   &lt;chr&gt;      &lt;chr&gt;     
 1 eXL30      C*04:01   
 2 eJL143     C*07:02   
 3 eXL36      B*15:01   
 4 eQD138     C*06:02:01
 5 eHO127     A*24:07:01
 6 eQD128     A*02:10   
 7 eQD138     A*01:01:01
 8 eMR21      C*07:02:01
 9 eHO133     A*33:01:01
10 ePD86      B*44:27:01</code></pre>
</div>
</div>
<p>Use the <code>View()</code>-function again, to look at the <code>meta_data</code> - Notice something? Some alleles are e.g.&nbsp;<code>A*11:01</code>, whereas others are <code>B*51:01:02</code>. You can find information on why, by visiting <a href="http://hla.alleles.org/nomenclature/naming.html">Nomenclature for Factors of the HLA System</a>.</p>
<p>Long story short, we only want to include <code>Field 1</code> (allele group) and <code>Field 2</code> (Specific HLA protein). You have prepared the <code>stringr</code>-package for today. See if you can find a way to reduce e.g.&nbsp;<code>B*51:01:02</code> to <code>B*51:01</code> and then create a new variable <code>Allele_F_1_2</code> accordingly, while also removing the <code>...x</code> (where <code>x</code> is a number) subscripts from the <code>Gene</code>-variable (It is an artifact from having the data in a wide format, where you cannot have two variables with the same name) and also, remove any <code>NA</code>s and <code>""</code>s, denoting empty entries.</p>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>There are several ways this can be achieved, the easiest being to consider if perhaps a part of the string based on indices could be of interest. This term “a part of a string” is called a substring, perhaps the <code>stringr</code>-package contains a function work with substring? In the console, type <code>stringr::</code> and hit <code>tab</code>. This will display the functions available in the <code>stringr</code>-package. Scroll down and find the functionst starting with <code>str_</code> and look for on, which might be relevant and remember you can use <code>?function_name</code> to get more information on how a given function works.</em>
</details>
<div class="cell">

</div>
<ul>
<li><strong>T19: Create the following data, according to specifications above:</strong></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   Experiment Allele     Allele_F_1_2
   &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;       
 1 ePD85      B*07:05    B*07:05     
 2 eAM23      A*24:02:01 A*24:02     
 3 eQD123     B*07:02:01 B*07:02     
 4 eEE226     B*35:02    B*35:02     
 5 eJL154     B*44:03:01 B*44:03     
 6 ePD80      B*41:02:01 B*41:02     
 7 eMR15      B*07:02:01 B*07:02     
 8 eLH47      B*07:02:01 B*07:02     
 9 eLH44      C*12:03:01 C*12:03     
10 eJL161     C*06:02:01 C*06:02     </code></pre>
</div>
</div>
<p>The asterix, i.e.&nbsp;<code>*</code> is a rather annoying character because of ambiguity, so:</p>
<ul>
<li><strong>T20: Clean the data a bit more, by removing the asterix and redundant variables:</strong></li>
</ul>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="sc">|&gt;</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="at">size =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 2
   Experiment Allele
   &lt;chr&gt;      &lt;chr&gt; 
 1 eAV105     A68:01
 2 eOX46      C04:01
 3 eDH96      C03:03
 4 eLH50      B15:01
 5 ePD80      A02:01
 6 eHO127     B40:02
 7 eLH47      A01:01
 8 eHH169     A74:01
 9 eHH174     A01:01
10 eAV105     C07:02</code></pre>
</div>
</div>
<details>
<p>
</p><summary>
Click here for hint 1
</summary>
<p></p>
<em>Again, the <code>stringr</code>-package may come in handy. Perhaps there is a function <code>remove</code>, one or more such pesky characters?</em>
</details>
<details>
<p>
</p><summary>
Click here for hint 2
</summary>
<p></p>
<em>Getting a weird error? Recall, that character ambiguity needs to be “escaped”, you did this somehow earlier on…</em>
</details>
<p>Recall the <code>peptide_data</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>peptide_data <span class="sc">|&gt;</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 7
   Experiment CDR3b             V_gene     J_gene     peptide  k_CDR3b k_peptide
   &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;      &lt;int&gt;     &lt;int&gt;
 1 eXL37      CASSGLAHRSDTQYF   TCRBV06-05 TCRBJ02-03 YTMADLV…      15         9
 2 ePD82      CASSQVASEFYEQYF   TCRBV04-01 TCRBJ02-07 ILGTVSW…      15         9
 3 eXL30      CASSAFEGDEQFF     TCRBV25-01 TCRBJ02-01 FPNITNL…      13        10
 4 eOX46      CASSSFSNQPQHF     TCRBV27-01 TCRBJ01-05 FTNVYAD…      13        10
 5 eOX49      CASSVVPGDYGYTF    TCRBV09-01 TCRBJ01-02 FLWLLWP…      14         9
 6 eXL30      CSADIVTNTEAFF     TCRBV20-X  TCRBJ01-01 AFLLFLV…      13         9
 7 eEE228     CSHLAKKYNEQFF     TCRBV29-01 TCRBJ02-01 AIPTNFT…      13        10
 8 eXL37      CSARGLAGGTNSYEQYF TCRBV20-X  TCRBJ02-07 IDFYLCF…      17        10
 9 eEE224     CASSLASLVGGTDTQYF TCRBV07-08 TCRBJ02-03 ATSRTLS…      17        10
10 eLH48      CASSIGWINGYTF     TCRBV19-01 TCRBJ01-02 YLYALVY…      13         9</code></pre>
</div>
</div>
<ul>
<li><strong>T21: Create a <code>dplyr</code>-pipeline, starting with the <code>peptide_data</code>, which joins it with the <code>meta_data</code> and remember to make sure that you get only unqiue observations of rows. Save this data into a new variable names <code>peptide_meta_data</code> (If you get a warning, discuss in your group what it means?)</strong></li>
</ul>
<div class="cell">

</div>
<details>
<p>
</p><summary>
Click here for hint 1
</summary>
<p></p>
<em>Which family of functions do we use to join data? Also, perhaps here it would be prudent to start with working on a smaller data set, recall we could sample a number of rows yielding a smaller development data set</em>
</details>
<details>
<p>
</p><summary>
Click here for hint 2
</summary>
<p></p>
<em>You should get a data set of around +3.000.000, <strong>take a moment to consider how that would have been to work with in Excel?</strong> Also, in case the servers are not liking this, you can consider subsetting the <code>peptide_data</code> prior to joining to e.g.&nbsp;100,000 or 10,000 rows.</em>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>peptide_meta_data <span class="sc">|&gt;</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 8
   Experiment CDR3b             V_gene   J_gene peptide k_CDR3b k_peptide Allele
   &lt;chr&gt;      &lt;chr&gt;             &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;chr&gt; 
 1 eOX56      CASSSGPDTQYF      TCRBV07… TCRBJ… WLLWPV…      12         9 A02:01
 2 eOX43      CSASGEVGDEQFF     TCRBV20… TCRBJ… SLIDFY…      13        10 C07:04
 3 eXL37      CASSRPRQGSTDTQYF  TCRBV07… TCRBJ… LYSPIF…      16         9 C16:01
 4 eLH47      CSVPLSTANTGELFF   TCRBV29… TCRBJ… AFLLFL…      15         9 C07:02
 5 eAV88      CSVNGWGSRSSYNEQFF TCRBV29… TCRBJ… FLWLLW…      17        10 A02:01
 6 eXL31      CASSQFDDEQYF      TCRBV04… TCRBJ… LWLLWP…      12         9 A02:01
 7 eXL31      CASSPSQEQYF       TCRBV28… TCRBJ… TLACFV…      11        10 A29:02
 8 eQD124     CASSYYNEQFF       TCRBV27… TCRBJ… HTTDPS…      11        11 C01:02
 9 eOX49      CASSKARTSGSGQTQYF TCRBV21… TCRBJ… LIDFYL…      17         9 B52:01
10 eEE228     CASSSYRDRGHEQYF   TCRBV27… TCRBJ… TESIVR…      15        10 C05:01</code></pre>
</div>
</div>
</section>
</section>
<section id="analysis" class="level3">
<h3 class="anchored" data-anchor-id="analysis">Analysis</h3>
<p>Now, that we have the data in a prepared and ready-to-analyse format, let us return to the two burning questions we had:</p>
<ol type="1">
<li>What characterises the peptides binding to the HLAs?</li>
<li>What characterises T-cell Receptors binding to the pMHC-complexes?</li>
</ol>
<section id="peptides-binding-to-hla" class="level4">
<h4 class="anchored" data-anchor-id="peptides-binding-to-hla">Peptides binding to HLA</h4>
<p>As we have touched upon multiple times, <code>R</code> is <em>very</em> flexible and naturally you can also create sequence logos. Finally, let us create a binding motif using the package <code>ggseqlogo</code> (<a href="https://omarwagih.github.io/ggseqlogo/">More info here</a>).</p>
<ul>
<li><strong>T22: Subset the final <code>peptide_meta_data</code>-data to <code>A02:01</code> and unique observations of peptides of length 9 and re-create the below sequence logo</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>You can pipe a vector of peptides into ggseqlogo, but perhaps you first need to <code>pull</code> that vector from the relevant variable in your tibble? Also, consider before that, that you’ll need to make sure, you are only looking at peptides of length 9</em>
</details>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="lab05_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>T23: Repeat for e.g.&nbsp;<code>B07:02</code> or another of your favorite alleles</strong></li>
</ul>
<p>Now, let’s take a closer look at the sequence logo:</p>
<ul>
<li><strong>Q14: Which positions in the peptide determines binding to HLA?</strong></li>
</ul>
<details>
<p>
</p><summary>
Click here for hint
</summary>
<p></p>
<em>Recall your Introduction to Bioinformatics course? And/or perhaps ask your fellow group members if they know?</em>
</details>
</section>
<section id="cdr3b-sequences-binding-to-pmhc" class="level4">
<h4 class="anchored" data-anchor-id="cdr3b-sequences-binding-to-pmhc">CDR3b-sequences binding to pMHC</h4>
<ul>
<li><strong>T24: Subset the <code>peptide_meta_data</code>, such that the length of the CDR3b is 15, the allele is A02:01 and the peptide is LLFLVLIML and re-create the below sequence logo of the CDR3b sequences:</strong></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="lab05_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><p><strong>Q15: In your group, discuss what you see?</strong></p></li>
<li><p><strong>T25: Play around with other combinations of <code>k_CDR3b</code>, <code>Allele</code>, and <code>peptide</code> and inspect how the logo changes</strong></p></li>
</ul>
<p><em>Disclaimer: In this data set, we only get: A given CDR3b was found to recognise a given peptide in a given subject and that subject had a given haplotype - Something’s missing… Perhaps if you have had immunology, then you can spot it? There is a trick to get around this missing information, but that’s beyond scope of what we’re working with here.</em></p>
</section>
</section>
</section>
<section id="epilogue" class="level2">
<h2 class="anchored" data-anchor-id="epilogue">Epilogue</h2>
<p>That’s it for today - I know this overwhelming now, but commit to it and you WILL be plenty rewarded! I hope today was at least a glimpse into the flexibility and capabilities of using <code>tidyverse</code> for applied Bio Data Science</p>
<p>…also, noticed something? We spend maybe 80% of the time here on dealing with data-wrangling and then once we’re good to go, the analysis wasn’t that time consuming - That’s often the way it ends up going, you’ll spend a lot of time on data handling and getting the tidyverse toolbox in your toolbelt, will allow you to be so much more effecient in your data wrangling, so you can get to the fun part as quick as possible!</p>
</section>
<section id="todays-assignment" class="level2">
<h2 class="anchored" data-anchor-id="todays-assignment"><span style="color: red;">Today’s Assignment</span></h2>
<p>Your group assignment today is to condense the exercises into a group micro-report! Talk together and figure out how to paste your exercises from today into one small end-to-end runable reproducible micro-report.</p>
<p>After today, we are halfway through the labs of the course, so now is a good time to spend some time recalling what we have been over and practising writing a reproducible Quarto-report.</p>
<p>Include the steps necessary to arrive at the final sequence logos, but be as concise as possible.</p>
<p><strong>IMPORTANT:</strong> Remember to check the <a href="https://r4bds.github.io/assignments.html">ASSIGNMENT GUIDELINES</a></p>
<p>…and as always - Have fun!</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./lab04.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Lab 4: Data Wrangling I</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./lab06.html" class="pagination-link">
        <span class="nav-page-text">Lab 6: Functional Programming Purrr</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">R for Bio Data Science</div>   
    <div class="nav-footer-right">Full course description available at <a href="https://kurser.dtu.dk/course/22100">DTU Course Base</a>.</div>
  </div>
</footer>



</body></html>